{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build the SDoH KG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### combine results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "df1 = pd.read_csv('result/server/resultgpt-4o.csv')\n",
    "df1.loc[-1] = df1.columns  # Add the column names as the first row\n",
    "df1.index = df1.index + 1  # Shift the index\n",
    "df1 = df1.sort_index()  # Sort the index to make the first row at the top\n",
    "df1.columns = ['Bio', 'Relationship', 'SDoH factor', 'identifier', 'Bio_type', 'source', 'pmid', 'sub_type', 'type', 'probability']\n",
    "\n",
    "df2 = pd.read_csv('result/server/coo (3).csv')\n",
    "df2.loc[-1] = df2.columns  # Add the column names as the first row\n",
    "df2.index = df2.index + 1  # Shift the index\n",
    "df2 = df2.sort_index()  # Sort the index to make the first row at the top\n",
    "df2.columns = ['Bio', 'Relationship', 'SDoH factor', 'identifier', 'Bio_type', 'source', 'pmid', 'sub_type', 'type', 'probability']\n",
    "\n",
    "df3 = pd.read_csv('result/test/resultgpt-4o.csv')\n",
    "df3.loc[-1] = df3.columns  # Add the column names as the first row\n",
    "df3.index = df3.index + 1  # Shift the index\n",
    "df3 = df3.sort_index()  # Sort the index to make the first row at the top\n",
    "df3.columns = ['Bio', 'Relationship', 'SDoH factor', 'identifier', 'Bio_type', 'source', 'pmid', 'sub_type', 'type', 'probability']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bio</th>\n",
       "      <th>Relationship</th>\n",
       "      <th>SDoH factor</th>\n",
       "      <th>identifier</th>\n",
       "      <th>Bio_type</th>\n",
       "      <th>source</th>\n",
       "      <th>pmid</th>\n",
       "      <th>sub_type</th>\n",
       "      <th>type</th>\n",
       "      <th>probability</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>patients</td>\n",
       "      <td>responsible</td>\n",
       "      <td>caregivers</td>\n",
       "      <td>9606</td>\n",
       "      <td>Species</td>\n",
       "      <td>ncbi_taxonomy</td>\n",
       "      <td>35484699</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>race</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>31601516</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>dementia</td>\n",
       "      <td>affects</td>\n",
       "      <td>families</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>34962459</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>0.428571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>dementia</td>\n",
       "      <td>affects</td>\n",
       "      <td>family member</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>34962459</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>0.428571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>diagnosis</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>35301255</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3291</th>\n",
       "      <td>heart disease</td>\n",
       "      <td>associates</td>\n",
       "      <td>poor nutrition</td>\n",
       "      <td>MESH:D006331</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>38304166</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3292</th>\n",
       "      <td>respiratory diseases</td>\n",
       "      <td>associates</td>\n",
       "      <td>poor nutrition</td>\n",
       "      <td>MESH:D012140</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>38304166</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3293</th>\n",
       "      <td>dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>disparities</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>38384925</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3294</th>\n",
       "      <td>Dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>representation</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>38460119</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3295</th>\n",
       "      <td>Alzheimer's disease</td>\n",
       "      <td>places burden on</td>\n",
       "      <td>families</td>\n",
       "      <td>MESH:D000544</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>38560360</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3296 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Bio      Relationship     SDoH factor    identifier  \\\n",
       "0                 patients       responsible      caregivers          9606   \n",
       "1                 dementia        associates            race  MESH:D003704   \n",
       "2                 dementia           affects        families  MESH:D003704   \n",
       "3                 dementia           affects   family member  MESH:D003704   \n",
       "4                 dementia        associates       diagnosis  MESH:D003704   \n",
       "...                    ...               ...             ...           ...   \n",
       "3291         heart disease        associates  poor nutrition  MESH:D006331   \n",
       "3292  respiratory diseases        associates  poor nutrition  MESH:D012140   \n",
       "3293              dementia        associates     disparities  MESH:D003704   \n",
       "3294              Dementia        associates  representation  MESH:D003704   \n",
       "3295   Alzheimer's disease  places burden on        families  MESH:D000544   \n",
       "\n",
       "     Bio_type         source      pmid  sub_type  type  probability  \n",
       "0     Species  ncbi_taxonomy  35484699         4     4     0.500000  \n",
       "1     Disease      ncbi_mesh  31601516         2     2     0.500000  \n",
       "2     Disease      ncbi_mesh  34962459         4     4     0.428571  \n",
       "3     Disease      ncbi_mesh  34962459         4     4     0.428571  \n",
       "4     Disease      ncbi_mesh  35301255         1     1     0.500000  \n",
       "...       ...            ...       ...       ...   ...          ...  \n",
       "3291  Disease      ncbi_mesh  38304166         1     1     0.500000  \n",
       "3292  Disease      ncbi_mesh  38304166         1     1     1.000000  \n",
       "3293  Disease      ncbi_mesh  38384925         2     2     1.000000  \n",
       "3294  Disease      ncbi_mesh  38460119         1     1     0.500000  \n",
       "3295  Disease      ncbi_mesh  38560360         4     4     1.000000  \n",
       "\n",
       "[3296 rows x 10 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result = pd.concat([df1, df2]).reset_index(drop=True)\n",
    "result = pd.concat([result, df3]).reset_index(drop=True)\n",
    "result['sub_type'] = result['sub_type'].astype(float).astype(int)\n",
    "result['type'] = result['sub_type'].astype(float).astype(int)\n",
    "result['probability'] = result['probability'].astype(float)\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bio</th>\n",
       "      <th>Relationship</th>\n",
       "      <th>SDoH factor</th>\n",
       "      <th>identifier</th>\n",
       "      <th>Bio_type</th>\n",
       "      <th>source</th>\n",
       "      <th>pmid</th>\n",
       "      <th>sub_type</th>\n",
       "      <th>type</th>\n",
       "      <th>probability</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>diagnosis</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>35301255</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>disease-modifying treatments</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>35301255</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>elderly care</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>35301255</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>guidelines</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>35301255</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>protocols</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>35301255</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3291</th>\n",
       "      <td>heart disease</td>\n",
       "      <td>associates</td>\n",
       "      <td>poor nutrition</td>\n",
       "      <td>MESH:D006331</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>38304166</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3292</th>\n",
       "      <td>respiratory diseases</td>\n",
       "      <td>associates</td>\n",
       "      <td>poor nutrition</td>\n",
       "      <td>MESH:D012140</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>38304166</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3293</th>\n",
       "      <td>dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>disparities</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>38384925</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3294</th>\n",
       "      <td>Dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>representation</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>38460119</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3295</th>\n",
       "      <td>Alzheimer's disease</td>\n",
       "      <td>places burden on</td>\n",
       "      <td>families</td>\n",
       "      <td>MESH:D000544</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>38560360</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2709 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Bio      Relationship                   SDoH factor  \\\n",
       "4                 dementia        associates                     diagnosis   \n",
       "5                 dementia        associates  disease-modifying treatments   \n",
       "6                 dementia        associates                  elderly care   \n",
       "7                 dementia        associates                    guidelines   \n",
       "8                 dementia        associates                     protocols   \n",
       "...                    ...               ...                           ...   \n",
       "3291         heart disease        associates                poor nutrition   \n",
       "3292  respiratory diseases        associates                poor nutrition   \n",
       "3293              dementia        associates                   disparities   \n",
       "3294              Dementia        associates                representation   \n",
       "3295   Alzheimer's disease  places burden on                      families   \n",
       "\n",
       "        identifier Bio_type     source      pmid  sub_type  type  probability  \n",
       "4     MESH:D003704  Disease  ncbi_mesh  35301255         1     1          0.5  \n",
       "5     MESH:D003704  Disease  ncbi_mesh  35301255         1     1          0.5  \n",
       "6     MESH:D003704  Disease  ncbi_mesh  35301255         1     1          0.5  \n",
       "7     MESH:D003704  Disease  ncbi_mesh  35301255         1     1          0.5  \n",
       "8     MESH:D003704  Disease  ncbi_mesh  35301255         1     1          0.5  \n",
       "...            ...      ...        ...       ...       ...   ...          ...  \n",
       "3291  MESH:D006331  Disease  ncbi_mesh  38304166         1     1          0.5  \n",
       "3292  MESH:D012140  Disease  ncbi_mesh  38304166         1     1          1.0  \n",
       "3293  MESH:D003704  Disease  ncbi_mesh  38384925         2     2          1.0  \n",
       "3294  MESH:D003704  Disease  ncbi_mesh  38460119         1     1          0.5  \n",
       "3295  MESH:D000544  Disease  ncbi_mesh  38560360         4     4          1.0  \n",
       "\n",
       "[2709 rows x 10 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for each row, if 'identifier' 'sub_type' 'type' are the same, keep the row with the highest 'probability'\n",
    "result = result.sort_values('probability', ascending=False).drop_duplicates(['identifier', 'SDoH factor']).sort_index() \n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "result.to_csv('result/resultgpt-4o.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bio</th>\n",
       "      <th>Relationship</th>\n",
       "      <th>SDoH factor</th>\n",
       "      <th>identifier</th>\n",
       "      <th>Bio_type</th>\n",
       "      <th>source</th>\n",
       "      <th>pmid</th>\n",
       "      <th>sub_type</th>\n",
       "      <th>type</th>\n",
       "      <th>probability</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>diagnosis</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>35301255</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>disease-modifying treatments</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>35301255</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>elderly care</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>35301255</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>guidelines</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>35301255</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>protocols</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>35301255</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2704</th>\n",
       "      <td>heart disease</td>\n",
       "      <td>associates</td>\n",
       "      <td>poor nutrition</td>\n",
       "      <td>MESH:D006331</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>38304166</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2705</th>\n",
       "      <td>respiratory diseases</td>\n",
       "      <td>associates</td>\n",
       "      <td>poor nutrition</td>\n",
       "      <td>MESH:D012140</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>38304166</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2706</th>\n",
       "      <td>dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>disparities</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>38384925</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2707</th>\n",
       "      <td>Dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>representation</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>38460119</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2708</th>\n",
       "      <td>Alzheimer's disease</td>\n",
       "      <td>places burden on</td>\n",
       "      <td>families</td>\n",
       "      <td>MESH:D000544</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>38560360</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2709 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Bio      Relationship                   SDoH factor  \\\n",
       "0                 dementia        associates                     diagnosis   \n",
       "1                 dementia        associates  disease-modifying treatments   \n",
       "2                 dementia        associates                  elderly care   \n",
       "3                 dementia        associates                    guidelines   \n",
       "4                 dementia        associates                     protocols   \n",
       "...                    ...               ...                           ...   \n",
       "2704         heart disease        associates                poor nutrition   \n",
       "2705  respiratory diseases        associates                poor nutrition   \n",
       "2706              dementia        associates                   disparities   \n",
       "2707              Dementia        associates                representation   \n",
       "2708   Alzheimer's disease  places burden on                      families   \n",
       "\n",
       "        identifier Bio_type     source      pmid  sub_type  type  probability  \n",
       "0     MESH:D003704  Disease  ncbi_mesh  35301255         1     1          0.5  \n",
       "1     MESH:D003704  Disease  ncbi_mesh  35301255         1     1          0.5  \n",
       "2     MESH:D003704  Disease  ncbi_mesh  35301255         1     1          0.5  \n",
       "3     MESH:D003704  Disease  ncbi_mesh  35301255         1     1          0.5  \n",
       "4     MESH:D003704  Disease  ncbi_mesh  35301255         1     1          0.5  \n",
       "...            ...      ...        ...       ...       ...   ...          ...  \n",
       "2704  MESH:D006331  Disease  ncbi_mesh  38304166         1     1          0.5  \n",
       "2705  MESH:D012140  Disease  ncbi_mesh  38304166         1     1          1.0  \n",
       "2706  MESH:D003704  Disease  ncbi_mesh  38384925         2     2          1.0  \n",
       "2707  MESH:D003704  Disease  ncbi_mesh  38460119         1     1          0.5  \n",
       "2708  MESH:D000544  Disease  ncbi_mesh  38560360         4     4          1.0  \n",
       "\n",
       "[2709 rows x 10 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SDoH_df = pd.read_csv('result/resultgpt-4o.csv')\n",
    "SDoH_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### map type and sub_type (int -> str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "type = ['healthcare access and quality', 'education access and quality', 'social and community context', 'economic stability', 'neighborhood and built environment']\n",
    "sub_type = [['Access to Health Services', 'Access to Primary Care', 'Health Literacy'],\n",
    "                  ['Early Childhood Development and Education', 'Enrollment in Higher Education', 'High School Graduation', 'Language and Literacy'],\n",
    "                  ['Civic Participation', 'Discrimination', 'Incarceration', 'Social Cohesion'],\n",
    "                  ['Employment', 'Food Insecurity', 'Housing Instability', 'Poverty'],\n",
    "                  ['Access to Foods That Support Healthy Dietary Patterns' ,'Crime and Violence' ,'Environmental Conditions' ,'Quality of Housing']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relation</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_source</th>\n",
       "      <th>y_name</th>\n",
       "      <th>y_id</th>\n",
       "      <th>y_type</th>\n",
       "      <th>y_source</th>\n",
       "      <th>pmid</th>\n",
       "      <th>probability</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>Access to Health Services</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>healthcare access and quality</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>Access to Primary Care</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>healthcare access and quality</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>Health Literacy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>healthcare access and quality</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>Early Childhood Development and Education</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>education access and quality</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>Enrollment in Higher Education</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>education access and quality</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>High School Graduation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>education access and quality</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>Language and Literacy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>education access and quality</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>Civic Participation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>social and community context</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>Discrimination</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>social and community context</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>Incarceration</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>social and community context</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>Social Cohesion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>social and community context</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>Employment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>economic stability</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>Food Insecurity</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>economic stability</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>Housing Instability</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>economic stability</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>Poverty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>economic stability</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>Access to Foods That Support Healthy Dietary P...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>neighborhood and built environment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>Crime and Violence</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>neighborhood and built environment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>Environmental Conditions</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>neighborhood and built environment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>Quality of Housing</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>neighborhood and built environment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       relation                                             x_name x_id  \\\n",
       "0   subordinate                          Access to Health Services  NaN   \n",
       "1   subordinate                             Access to Primary Care  NaN   \n",
       "2   subordinate                                    Health Literacy  NaN   \n",
       "3   subordinate          Early Childhood Development and Education  NaN   \n",
       "4   subordinate                     Enrollment in Higher Education  NaN   \n",
       "5   subordinate                             High School Graduation  NaN   \n",
       "6   subordinate                              Language and Literacy  NaN   \n",
       "7   subordinate                                Civic Participation  NaN   \n",
       "8   subordinate                                     Discrimination  NaN   \n",
       "9   subordinate                                      Incarceration  NaN   \n",
       "10  subordinate                                    Social Cohesion  NaN   \n",
       "11  subordinate                                         Employment  NaN   \n",
       "12  subordinate                                    Food Insecurity  NaN   \n",
       "13  subordinate                                Housing Instability  NaN   \n",
       "14  subordinate                                            Poverty  NaN   \n",
       "15  subordinate  Access to Foods That Support Healthy Dietary P...  NaN   \n",
       "16  subordinate                                 Crime and Violence  NaN   \n",
       "17  subordinate                           Environmental Conditions  NaN   \n",
       "18  subordinate                                 Quality of Housing  NaN   \n",
       "\n",
       "          x_type       x_source                              y_name y_id  \\\n",
       "0   SDoH subtype  healthypeople       healthcare access and quality  NaN   \n",
       "1   SDoH subtype  healthypeople       healthcare access and quality  NaN   \n",
       "2   SDoH subtype  healthypeople       healthcare access and quality  NaN   \n",
       "3   SDoH subtype  healthypeople        education access and quality  NaN   \n",
       "4   SDoH subtype  healthypeople        education access and quality  NaN   \n",
       "5   SDoH subtype  healthypeople        education access and quality  NaN   \n",
       "6   SDoH subtype  healthypeople        education access and quality  NaN   \n",
       "7   SDoH subtype  healthypeople        social and community context  NaN   \n",
       "8   SDoH subtype  healthypeople        social and community context  NaN   \n",
       "9   SDoH subtype  healthypeople        social and community context  NaN   \n",
       "10  SDoH subtype  healthypeople        social and community context  NaN   \n",
       "11  SDoH subtype  healthypeople                  economic stability  NaN   \n",
       "12  SDoH subtype  healthypeople                  economic stability  NaN   \n",
       "13  SDoH subtype  healthypeople                  economic stability  NaN   \n",
       "14  SDoH subtype  healthypeople                  economic stability  NaN   \n",
       "15  SDoH subtype  healthypeople  neighborhood and built environment  NaN   \n",
       "16  SDoH subtype  healthypeople  neighborhood and built environment  NaN   \n",
       "17  SDoH subtype  healthypeople  neighborhood and built environment  NaN   \n",
       "18  SDoH subtype  healthypeople  neighborhood and built environment  NaN   \n",
       "\n",
       "       y_type       y_source pmid  probability  \n",
       "0   SDoH type  healthypeople  NaN          1.0  \n",
       "1   SDoH type  healthypeople  NaN          1.0  \n",
       "2   SDoH type  healthypeople  NaN          1.0  \n",
       "3   SDoH type  healthypeople  NaN          1.0  \n",
       "4   SDoH type  healthypeople  NaN          1.0  \n",
       "5   SDoH type  healthypeople  NaN          1.0  \n",
       "6   SDoH type  healthypeople  NaN          1.0  \n",
       "7   SDoH type  healthypeople  NaN          1.0  \n",
       "8   SDoH type  healthypeople  NaN          1.0  \n",
       "9   SDoH type  healthypeople  NaN          1.0  \n",
       "10  SDoH type  healthypeople  NaN          1.0  \n",
       "11  SDoH type  healthypeople  NaN          1.0  \n",
       "12  SDoH type  healthypeople  NaN          1.0  \n",
       "13  SDoH type  healthypeople  NaN          1.0  \n",
       "14  SDoH type  healthypeople  NaN          1.0  \n",
       "15  SDoH type  healthypeople  NaN          1.0  \n",
       "16  SDoH type  healthypeople  NaN          1.0  \n",
       "17  SDoH type  healthypeople  NaN          1.0  \n",
       "18  SDoH type  healthypeople  NaN          1.0  "
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a list to store the rows\n",
    "rows = []\n",
    "\n",
    "# Loop through types and their corresponding sub_types\n",
    "for type_value, sub_type_list in zip(type, sub_type):\n",
    "    for sub_type_value in sub_type_list:\n",
    "        rows.append({'relation': 'subordinate', 'x_name': sub_type_value, 'x_id': 'NaN', 'x_type':'SDoH subtype', 'x_source':'healthypeople',\n",
    "                     'y_name': type_value, 'y_id': 'NaN', 'y_type':'SDoH type', 'y_source':'healthypeople', 'pmid': 'NaN', 'probability': 1.0})\n",
    "# Create the DataFrame\n",
    "typedf = pd.DataFrame(rows)\n",
    "typedf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bio</th>\n",
       "      <th>Relationship</th>\n",
       "      <th>SDoH factor</th>\n",
       "      <th>identifier</th>\n",
       "      <th>Bio_type</th>\n",
       "      <th>source</th>\n",
       "      <th>pmid</th>\n",
       "      <th>sub_type</th>\n",
       "      <th>type</th>\n",
       "      <th>probability</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>diagnosis</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>35301255</td>\n",
       "      <td>Access to Health Services</td>\n",
       "      <td>healthcare access and quality</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>disease-modifying treatments</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>35301255</td>\n",
       "      <td>Access to Health Services</td>\n",
       "      <td>healthcare access and quality</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>elderly care</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>35301255</td>\n",
       "      <td>Access to Health Services</td>\n",
       "      <td>healthcare access and quality</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>guidelines</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>35301255</td>\n",
       "      <td>Access to Health Services</td>\n",
       "      <td>healthcare access and quality</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>protocols</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>35301255</td>\n",
       "      <td>Access to Health Services</td>\n",
       "      <td>healthcare access and quality</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2704</th>\n",
       "      <td>heart disease</td>\n",
       "      <td>associates</td>\n",
       "      <td>poor nutrition</td>\n",
       "      <td>MESH:D006331</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>38304166</td>\n",
       "      <td>Access to Health Services</td>\n",
       "      <td>healthcare access and quality</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2705</th>\n",
       "      <td>respiratory diseases</td>\n",
       "      <td>associates</td>\n",
       "      <td>poor nutrition</td>\n",
       "      <td>MESH:D012140</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>38304166</td>\n",
       "      <td>Access to Health Services</td>\n",
       "      <td>healthcare access and quality</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2706</th>\n",
       "      <td>dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>disparities</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>38384925</td>\n",
       "      <td>Enrollment in Higher Education</td>\n",
       "      <td>education access and quality</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2707</th>\n",
       "      <td>Dementia</td>\n",
       "      <td>associates</td>\n",
       "      <td>representation</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>38460119</td>\n",
       "      <td>Access to Health Services</td>\n",
       "      <td>healthcare access and quality</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2708</th>\n",
       "      <td>Alzheimer's disease</td>\n",
       "      <td>places burden on</td>\n",
       "      <td>families</td>\n",
       "      <td>MESH:D000544</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>38560360</td>\n",
       "      <td>Poverty</td>\n",
       "      <td>economic stability</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2709 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Bio      Relationship                   SDoH factor  \\\n",
       "0                 dementia        associates                     diagnosis   \n",
       "1                 dementia        associates  disease-modifying treatments   \n",
       "2                 dementia        associates                  elderly care   \n",
       "3                 dementia        associates                    guidelines   \n",
       "4                 dementia        associates                     protocols   \n",
       "...                    ...               ...                           ...   \n",
       "2704         heart disease        associates                poor nutrition   \n",
       "2705  respiratory diseases        associates                poor nutrition   \n",
       "2706              dementia        associates                   disparities   \n",
       "2707              Dementia        associates                representation   \n",
       "2708   Alzheimer's disease  places burden on                      families   \n",
       "\n",
       "        identifier Bio_type     source      pmid  \\\n",
       "0     MESH:D003704  Disease  ncbi_mesh  35301255   \n",
       "1     MESH:D003704  Disease  ncbi_mesh  35301255   \n",
       "2     MESH:D003704  Disease  ncbi_mesh  35301255   \n",
       "3     MESH:D003704  Disease  ncbi_mesh  35301255   \n",
       "4     MESH:D003704  Disease  ncbi_mesh  35301255   \n",
       "...            ...      ...        ...       ...   \n",
       "2704  MESH:D006331  Disease  ncbi_mesh  38304166   \n",
       "2705  MESH:D012140  Disease  ncbi_mesh  38304166   \n",
       "2706  MESH:D003704  Disease  ncbi_mesh  38384925   \n",
       "2707  MESH:D003704  Disease  ncbi_mesh  38460119   \n",
       "2708  MESH:D000544  Disease  ncbi_mesh  38560360   \n",
       "\n",
       "                            sub_type                           type  \\\n",
       "0          Access to Health Services  healthcare access and quality   \n",
       "1          Access to Health Services  healthcare access and quality   \n",
       "2          Access to Health Services  healthcare access and quality   \n",
       "3          Access to Health Services  healthcare access and quality   \n",
       "4          Access to Health Services  healthcare access and quality   \n",
       "...                              ...                            ...   \n",
       "2704       Access to Health Services  healthcare access and quality   \n",
       "2705       Access to Health Services  healthcare access and quality   \n",
       "2706  Enrollment in Higher Education   education access and quality   \n",
       "2707       Access to Health Services  healthcare access and quality   \n",
       "2708                         Poverty             economic stability   \n",
       "\n",
       "      probability  \n",
       "0             0.5  \n",
       "1             0.5  \n",
       "2             0.5  \n",
       "3             0.5  \n",
       "4             0.5  \n",
       "...           ...  \n",
       "2704          0.5  \n",
       "2705          1.0  \n",
       "2706          1.0  \n",
       "2707          0.5  \n",
       "2708          1.0  \n",
       "\n",
       "[2709 rows x 10 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for each row in SDoH_df, change the 'type' and 'sub_type' to the corresponding values\n",
    "for i in range(len(SDoH_df)):\n",
    "    SDoH_df.loc[i, 'sub_type'] = sub_type[int(SDoH_df.loc[i, 'type'])-1][int(SDoH_df.loc[i, 'sub_type'])-1] \n",
    "    SDoH_df.loc[i, 'type'] = type[int(SDoH_df.loc[i, 'type'])-1]\n",
    "\n",
    "SDoH_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "break [bio-sdoh-sub_type-type] to \n",
    "1. bio-sdoh (SDoHKG1)\n",
    "2. sdoh-sub_type (SDoHKG2)\n",
    "3. sub_type-type (typedf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relation</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_source</th>\n",
       "      <th>y_name</th>\n",
       "      <th>y_id</th>\n",
       "      <th>y_type</th>\n",
       "      <th>y_source</th>\n",
       "      <th>pmid</th>\n",
       "      <th>probability</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>associates</td>\n",
       "      <td>dementia</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>diagnosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>associates</td>\n",
       "      <td>dementia</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>disease-modifying treatments</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>associates</td>\n",
       "      <td>dementia</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>elderly care</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>associates</td>\n",
       "      <td>dementia</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>guidelines</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>associates</td>\n",
       "      <td>dementia</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>protocols</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2704</th>\n",
       "      <td>associates</td>\n",
       "      <td>heart disease</td>\n",
       "      <td>MESH:D006331</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>poor nutrition</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>38304166</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2705</th>\n",
       "      <td>associates</td>\n",
       "      <td>respiratory diseases</td>\n",
       "      <td>MESH:D012140</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>poor nutrition</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>38304166</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2706</th>\n",
       "      <td>associates</td>\n",
       "      <td>dementia</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>disparities</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>38384925</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2707</th>\n",
       "      <td>associates</td>\n",
       "      <td>Dementia</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>representation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>38460119</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2708</th>\n",
       "      <td>places burden on</td>\n",
       "      <td>Alzheimer's disease</td>\n",
       "      <td>MESH:D000544</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>families</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>38560360</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2709 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              relation                x_name          x_id   x_type  \\\n",
       "0           associates              dementia  MESH:D003704  Disease   \n",
       "1           associates              dementia  MESH:D003704  Disease   \n",
       "2           associates              dementia  MESH:D003704  Disease   \n",
       "3           associates              dementia  MESH:D003704  Disease   \n",
       "4           associates              dementia  MESH:D003704  Disease   \n",
       "...                ...                   ...           ...      ...   \n",
       "2704        associates         heart disease  MESH:D006331  Disease   \n",
       "2705        associates  respiratory diseases  MESH:D012140  Disease   \n",
       "2706        associates              dementia  MESH:D003704  Disease   \n",
       "2707        associates              Dementia  MESH:D003704  Disease   \n",
       "2708  places burden on   Alzheimer's disease  MESH:D000544  Disease   \n",
       "\n",
       "       x_source                        y_name y_id       y_type  \\\n",
       "0     ncbi_mesh                     diagnosis  NaN  SDoH factor   \n",
       "1     ncbi_mesh  disease-modifying treatments  NaN  SDoH factor   \n",
       "2     ncbi_mesh                  elderly care  NaN  SDoH factor   \n",
       "3     ncbi_mesh                    guidelines  NaN  SDoH factor   \n",
       "4     ncbi_mesh                     protocols  NaN  SDoH factor   \n",
       "...         ...                           ...  ...          ...   \n",
       "2704  ncbi_mesh                poor nutrition  NaN  SDoH factor   \n",
       "2705  ncbi_mesh                poor nutrition  NaN  SDoH factor   \n",
       "2706  ncbi_mesh                   disparities  NaN  SDoH factor   \n",
       "2707  ncbi_mesh                representation  NaN  SDoH factor   \n",
       "2708  ncbi_mesh                      families  NaN  SDoH factor   \n",
       "\n",
       "           y_source      pmid  probability  \n",
       "0     healthypeople  35301255          0.5  \n",
       "1     healthypeople  35301255          0.5  \n",
       "2     healthypeople  35301255          0.5  \n",
       "3     healthypeople  35301255          0.5  \n",
       "4     healthypeople  35301255          0.5  \n",
       "...             ...       ...          ...  \n",
       "2704  healthypeople  38304166          0.5  \n",
       "2705  healthypeople  38304166          1.0  \n",
       "2706  healthypeople  38384925          1.0  \n",
       "2707  healthypeople  38460119          0.5  \n",
       "2708  healthypeople  38560360          1.0  \n",
       "\n",
       "[2709 rows x 11 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SDoHKG1 = SDoH_df[['Relationship', 'Bio', 'identifier', 'Bio_type', 'source', 'SDoH factor', 'pmid', 'probability' ]].copy()\n",
    "#rename the columns\n",
    "SDoHKG1.columns = ['relation', 'x_name', 'x_id', 'x_type', 'x_source', 'y_name', 'pmid', 'probability']\n",
    "# SDoHKG1.loc[:, 'y_id'] = 'NaN'\n",
    "# SDoHKG1.loc[:, 'y_type'] = 'SDoH factor'\n",
    "# SDoHKG1.loc[:, 'y_source'] = 'NaN'\n",
    "SDoHKG1['y_id'] = 'NaN'\n",
    "SDoHKG1['y_type'] = 'SDoH factor'\n",
    "SDoHKG1['y_source'] = 'healthypeople'\n",
    "# move the column 'y_id' to the left of 'probability'\n",
    "SDoHKG1 = SDoHKG1[['relation', 'x_name', 'x_id', 'x_type', 'x_source', 'y_name', 'y_id', 'y_type', 'y_source', 'pmid', 'probability' ]]\n",
    "SDoHKG1 = SDoHKG1.drop_duplicates(subset=['x_id', 'y_name'], keep='first')\n",
    "SDoHKG1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relation</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_source</th>\n",
       "      <th>y_name</th>\n",
       "      <th>y_id</th>\n",
       "      <th>y_type</th>\n",
       "      <th>y_source</th>\n",
       "      <th>pmid</th>\n",
       "      <th>probability</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>diagnosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>Access to Health Services</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>disease-modifying treatments</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>Access to Health Services</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>elderly care</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>Access to Health Services</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>guidelines</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>Access to Health Services</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>protocols</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>Access to Health Services</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2699</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>non-hispanic black</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>Enrollment in Higher Education</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>37984835</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2701</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>utility hardship</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>Poverty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>38087286</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2702</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>income segregation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>Poverty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>38217979</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2703</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>poor nutrition</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>Access to Health Services</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>38304166</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2707</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>representation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>Access to Health Services</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>38460119</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1330 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         relation                        x_name x_id       x_type  \\\n",
       "0     subordinate                     diagnosis  NaN  SDoH factor   \n",
       "1     subordinate  disease-modifying treatments  NaN  SDoH factor   \n",
       "2     subordinate                  elderly care  NaN  SDoH factor   \n",
       "3     subordinate                    guidelines  NaN  SDoH factor   \n",
       "4     subordinate                     protocols  NaN  SDoH factor   \n",
       "...           ...                           ...  ...          ...   \n",
       "2699  subordinate            non-hispanic black  NaN  SDoH factor   \n",
       "2701  subordinate              utility hardship  NaN  SDoH factor   \n",
       "2702  subordinate            income segregation  NaN  SDoH factor   \n",
       "2703  subordinate                poor nutrition  NaN  SDoH factor   \n",
       "2707  subordinate                representation  NaN  SDoH factor   \n",
       "\n",
       "           x_source                          y_name y_id        y_type  \\\n",
       "0     healthypeople       Access to Health Services  NaN  SDoH subtype   \n",
       "1     healthypeople       Access to Health Services  NaN  SDoH subtype   \n",
       "2     healthypeople       Access to Health Services  NaN  SDoH subtype   \n",
       "3     healthypeople       Access to Health Services  NaN  SDoH subtype   \n",
       "4     healthypeople       Access to Health Services  NaN  SDoH subtype   \n",
       "...             ...                             ...  ...           ...   \n",
       "2699  healthypeople  Enrollment in Higher Education  NaN  SDoH subtype   \n",
       "2701  healthypeople                         Poverty  NaN  SDoH subtype   \n",
       "2702  healthypeople                         Poverty  NaN  SDoH subtype   \n",
       "2703  healthypeople       Access to Health Services  NaN  SDoH subtype   \n",
       "2707  healthypeople       Access to Health Services  NaN  SDoH subtype   \n",
       "\n",
       "           y_source      pmid  probability  \n",
       "0     healthypeople  35301255          1.0  \n",
       "1     healthypeople  35301255          1.0  \n",
       "2     healthypeople  35301255          1.0  \n",
       "3     healthypeople  35301255          1.0  \n",
       "4     healthypeople  35301255          1.0  \n",
       "...             ...       ...          ...  \n",
       "2699  healthypeople  37984835          1.0  \n",
       "2701  healthypeople  38087286          1.0  \n",
       "2702  healthypeople  38217979          1.0  \n",
       "2703  healthypeople  38304166          1.0  \n",
       "2707  healthypeople  38460119          1.0  \n",
       "\n",
       "[1330 rows x 11 columns]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SDoHKG2 = SDoH_df[['Relationship', 'SDoH factor', 'sub_type', 'pmid', 'probability']].copy()\n",
    "#change relationship values to to subordinate, probability to 1\n",
    "SDoHKG2.loc[:, 'Relationship'] = 'subordinate'\n",
    "SDoHKG2.loc[:, 'probability'] = 1.0\n",
    "SDoHKG2.loc[:, 'x_id'] = 'NaN'\n",
    "SDoHKG2.loc[:, 'x_type'] = 'SDoH factor'\n",
    "SDoHKG2.loc[:, 'x_source'] = 'healthypeople'\n",
    "SDoHKG2.loc[:, 'y_id'] = 'NaN'\n",
    "SDoHKG2.loc[:, 'y_type'] = 'SDoH subtype'\n",
    "SDoHKG2.loc[:, 'y_source'] = 'healthypeople'\n",
    "#rename the columns\n",
    "SDoHKG2.columns = ['relation', 'x_name', 'y_name', 'pmid', 'probability', 'x_id', 'x_type', 'x_source',  'y_id', 'y_type', 'y_source']\n",
    "SDoHKG2 = SDoHKG2[['relation', 'x_name', 'x_id', 'x_type', 'x_source', 'y_name', 'y_id', 'y_type', 'y_source', 'pmid', 'probability' ]]\n",
    "SDoHKG2['x_name'] = SDoHKG2['x_name'].str.lower()\n",
    "SDoHKG2 = SDoHKG2.drop_duplicates(subset=['x_name', 'y_name'], keep='first')\n",
    "SDoHKG2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "concat the 3 subgraphs, set the index to -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relation</th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_source</th>\n",
       "      <th>y_index</th>\n",
       "      <th>y_name</th>\n",
       "      <th>y_id</th>\n",
       "      <th>y_type</th>\n",
       "      <th>y_source</th>\n",
       "      <th>pmid</th>\n",
       "      <th>probability</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>associates</td>\n",
       "      <td>-1</td>\n",
       "      <td>dementia</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>-1</td>\n",
       "      <td>diagnosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>associates</td>\n",
       "      <td>-1</td>\n",
       "      <td>dementia</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>-1</td>\n",
       "      <td>disease-modifying treatments</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>associates</td>\n",
       "      <td>-1</td>\n",
       "      <td>dementia</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>-1</td>\n",
       "      <td>elderly care</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>associates</td>\n",
       "      <td>-1</td>\n",
       "      <td>dementia</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>-1</td>\n",
       "      <td>guidelines</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>associates</td>\n",
       "      <td>-1</td>\n",
       "      <td>dementia</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>-1</td>\n",
       "      <td>protocols</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4053</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>-1</td>\n",
       "      <td>Poverty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>-1</td>\n",
       "      <td>economic stability</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4054</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>-1</td>\n",
       "      <td>Access to Foods That Support Healthy Dietary P...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>-1</td>\n",
       "      <td>neighborhood and built environment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4055</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>-1</td>\n",
       "      <td>Crime and Violence</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>-1</td>\n",
       "      <td>neighborhood and built environment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4056</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>-1</td>\n",
       "      <td>Environmental Conditions</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>-1</td>\n",
       "      <td>neighborhood and built environment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4057</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>-1</td>\n",
       "      <td>Quality of Housing</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>-1</td>\n",
       "      <td>neighborhood and built environment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4058 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         relation  x_index                                             x_name  \\\n",
       "0      associates       -1                                           dementia   \n",
       "1      associates       -1                                           dementia   \n",
       "2      associates       -1                                           dementia   \n",
       "3      associates       -1                                           dementia   \n",
       "4      associates       -1                                           dementia   \n",
       "...           ...      ...                                                ...   \n",
       "4053  subordinate       -1                                            Poverty   \n",
       "4054  subordinate       -1  Access to Foods That Support Healthy Dietary P...   \n",
       "4055  subordinate       -1                                 Crime and Violence   \n",
       "4056  subordinate       -1                           Environmental Conditions   \n",
       "4057  subordinate       -1                                 Quality of Housing   \n",
       "\n",
       "              x_id        x_type       x_source  y_index  \\\n",
       "0     MESH:D003704       Disease      ncbi_mesh       -1   \n",
       "1     MESH:D003704       Disease      ncbi_mesh       -1   \n",
       "2     MESH:D003704       Disease      ncbi_mesh       -1   \n",
       "3     MESH:D003704       Disease      ncbi_mesh       -1   \n",
       "4     MESH:D003704       Disease      ncbi_mesh       -1   \n",
       "...            ...           ...            ...      ...   \n",
       "4053           NaN  SDoH subtype  healthypeople       -1   \n",
       "4054           NaN  SDoH subtype  healthypeople       -1   \n",
       "4055           NaN  SDoH subtype  healthypeople       -1   \n",
       "4056           NaN  SDoH subtype  healthypeople       -1   \n",
       "4057           NaN  SDoH subtype  healthypeople       -1   \n",
       "\n",
       "                                  y_name y_id       y_type       y_source  \\\n",
       "0                              diagnosis  NaN  SDoH factor  healthypeople   \n",
       "1           disease-modifying treatments  NaN  SDoH factor  healthypeople   \n",
       "2                           elderly care  NaN  SDoH factor  healthypeople   \n",
       "3                             guidelines  NaN  SDoH factor  healthypeople   \n",
       "4                              protocols  NaN  SDoH factor  healthypeople   \n",
       "...                                  ...  ...          ...            ...   \n",
       "4053                  economic stability  NaN    SDoH type  healthypeople   \n",
       "4054  neighborhood and built environment  NaN    SDoH type  healthypeople   \n",
       "4055  neighborhood and built environment  NaN    SDoH type  healthypeople   \n",
       "4056  neighborhood and built environment  NaN    SDoH type  healthypeople   \n",
       "4057  neighborhood and built environment  NaN    SDoH type  healthypeople   \n",
       "\n",
       "          pmid  probability  \n",
       "0     35301255          0.5  \n",
       "1     35301255          0.5  \n",
       "2     35301255          0.5  \n",
       "3     35301255          0.5  \n",
       "4     35301255          0.5  \n",
       "...        ...          ...  \n",
       "4053       NaN          1.0  \n",
       "4054       NaN          1.0  \n",
       "4055       NaN          1.0  \n",
       "4056       NaN          1.0  \n",
       "4057       NaN          1.0  \n",
       "\n",
       "[4058 rows x 13 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#merge SDoHKG and SDoHKG1\n",
    "SDoHKG = pd.concat([SDoHKG1, SDoHKG2], axis=0).reset_index(drop=True)\n",
    "#merge SDoHKG and typedf\n",
    "SDoHKG = pd.concat([SDoHKG, typedf], axis=0).reset_index(drop=True)\n",
    "SDoHKG['x_index'] = -1\n",
    "SDoHKG['y_index'] = -1\n",
    "# Step 2: Reorder the columns to place 'index' as the second column\n",
    "cols = SDoHKG.columns.tolist()\n",
    "cols.insert(1, cols.pop(cols.index('x_index')))\n",
    "cols.insert(6, cols.pop(cols.index('y_index')))\n",
    "# drop duplicate\n",
    "#SDoHKG.drop_duplicates(inplace=True)\n",
    "\n",
    "# reset index\n",
    "SDoHKG.reset_index(drop=True, inplace=True)\n",
    "SDoHKG = SDoHKG[cols]\n",
    "SDoHKG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "SDoHKG.to_csv('data\\SDoHKG.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Entitiy mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "get the entities of primeKG: pkg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relation</th>\n",
       "      <th>display_relation</th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_source</th>\n",
       "      <th>y_index</th>\n",
       "      <th>y_id</th>\n",
       "      <th>y_type</th>\n",
       "      <th>y_name</th>\n",
       "      <th>y_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>0</td>\n",
       "      <td>9796</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>PHYHIP</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>8889</td>\n",
       "      <td>56992</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>KIF15</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>1</td>\n",
       "      <td>7918</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>GPANK1</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>2798</td>\n",
       "      <td>9240</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>PNMA1</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>2</td>\n",
       "      <td>8233</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>5646</td>\n",
       "      <td>23548</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>TTC33</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>3</td>\n",
       "      <td>4899</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NRF1</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>11592</td>\n",
       "      <td>11253</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>MAN1B1</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>4</td>\n",
       "      <td>5297</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>PI4KA</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>2122</td>\n",
       "      <td>8601</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>RGS20</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8100493</th>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>expression absent</td>\n",
       "      <td>66747</td>\n",
       "      <td>4720</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>cerebellar vermis</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>5259</td>\n",
       "      <td>140</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>ADORA3</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8100494</th>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>expression absent</td>\n",
       "      <td>63824</td>\n",
       "      <td>1377</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>quadriceps femoris</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>58254</td>\n",
       "      <td>105378952</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>KLF18</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8100495</th>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>expression absent</td>\n",
       "      <td>63826</td>\n",
       "      <td>1379</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>vastus lateralis</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>58254</td>\n",
       "      <td>105378952</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>KLF18</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8100496</th>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>expression absent</td>\n",
       "      <td>64523</td>\n",
       "      <td>2084</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>heart left ventricle</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>58254</td>\n",
       "      <td>105378952</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>KLF18</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8100497</th>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>expression absent</td>\n",
       "      <td>67302</td>\n",
       "      <td>5384</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>nasal cavity epithelium</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>58254</td>\n",
       "      <td>105378952</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>KLF18</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8100498 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       relation   display_relation  x_index  x_id  \\\n",
       "0               protein_protein                ppi        0  9796   \n",
       "1               protein_protein                ppi        1  7918   \n",
       "2               protein_protein                ppi        2  8233   \n",
       "3               protein_protein                ppi        3  4899   \n",
       "4               protein_protein                ppi        4  5297   \n",
       "...                         ...                ...      ...   ...   \n",
       "8100493  anatomy_protein_absent  expression absent    66747  4720   \n",
       "8100494  anatomy_protein_absent  expression absent    63824  1377   \n",
       "8100495  anatomy_protein_absent  expression absent    63826  1379   \n",
       "8100496  anatomy_protein_absent  expression absent    64523  2084   \n",
       "8100497  anatomy_protein_absent  expression absent    67302  5384   \n",
       "\n",
       "               x_type                   x_name x_source  y_index       y_id  \\\n",
       "0        gene/protein                   PHYHIP     NCBI     8889      56992   \n",
       "1        gene/protein                   GPANK1     NCBI     2798       9240   \n",
       "2        gene/protein                    ZRSR2     NCBI     5646      23548   \n",
       "3        gene/protein                     NRF1     NCBI    11592      11253   \n",
       "4        gene/protein                    PI4KA     NCBI     2122       8601   \n",
       "...               ...                      ...      ...      ...        ...   \n",
       "8100493       anatomy        cerebellar vermis   UBERON     5259        140   \n",
       "8100494       anatomy       quadriceps femoris   UBERON    58254  105378952   \n",
       "8100495       anatomy         vastus lateralis   UBERON    58254  105378952   \n",
       "8100496       anatomy     heart left ventricle   UBERON    58254  105378952   \n",
       "8100497       anatomy  nasal cavity epithelium   UBERON    58254  105378952   \n",
       "\n",
       "               y_type  y_name y_source  \n",
       "0        gene/protein   KIF15     NCBI  \n",
       "1        gene/protein   PNMA1     NCBI  \n",
       "2        gene/protein   TTC33     NCBI  \n",
       "3        gene/protein  MAN1B1     NCBI  \n",
       "4        gene/protein   RGS20     NCBI  \n",
       "...               ...     ...      ...  \n",
       "8100493  gene/protein  ADORA3     NCBI  \n",
       "8100494  gene/protein   KLF18     NCBI  \n",
       "8100495  gene/protein   KLF18     NCBI  \n",
       "8100496  gene/protein   KLF18     NCBI  \n",
       "8100497  gene/protein   KLF18     NCBI  \n",
       "\n",
       "[8100498 rows x 12 columns]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "primeKG = pd.read_csv('data\\kg.csv', low_memory=False)\n",
    "primeKG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>PHYHIP</td>\n",
       "      <td>9796</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>GPANK1</td>\n",
       "      <td>7918</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>8233</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>NRF1</td>\n",
       "      <td>4899</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>PI4KA</td>\n",
       "      <td>5297</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129370</th>\n",
       "      <td>129370</td>\n",
       "      <td>Ion transport by P-type ATPases</td>\n",
       "      <td>R-HSA-936837</td>\n",
       "      <td>pathway</td>\n",
       "      <td>REACTOME</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129371</th>\n",
       "      <td>129371</td>\n",
       "      <td>Inhibition  of voltage gated Ca2+ channels via...</td>\n",
       "      <td>R-HSA-997272</td>\n",
       "      <td>pathway</td>\n",
       "      <td>REACTOME</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129372</th>\n",
       "      <td>129372</td>\n",
       "      <td>anatomical entity</td>\n",
       "      <td>1062</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>UBERON</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129373</th>\n",
       "      <td>129373</td>\n",
       "      <td>multi-cellular organism</td>\n",
       "      <td>468</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>UBERON</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129374</th>\n",
       "      <td>129374</td>\n",
       "      <td>female gonad</td>\n",
       "      <td>992</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>UBERON</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>129375 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        x_index                                             x_name  \\\n",
       "0             0                                             PHYHIP   \n",
       "1             1                                             GPANK1   \n",
       "2             2                                              ZRSR2   \n",
       "3             3                                               NRF1   \n",
       "4             4                                              PI4KA   \n",
       "...         ...                                                ...   \n",
       "129370   129370                    Ion transport by P-type ATPases   \n",
       "129371   129371  Inhibition  of voltage gated Ca2+ channels via...   \n",
       "129372   129372                                  anatomical entity   \n",
       "129373   129373                            multi-cellular organism   \n",
       "129374   129374                                       female gonad   \n",
       "\n",
       "                x_id        x_type  x_source  \n",
       "0               9796  gene/protein      NCBI  \n",
       "1               7918  gene/protein      NCBI  \n",
       "2               8233  gene/protein      NCBI  \n",
       "3               4899  gene/protein      NCBI  \n",
       "4               5297  gene/protein      NCBI  \n",
       "...              ...           ...       ...  \n",
       "129370  R-HSA-936837       pathway  REACTOME  \n",
       "129371  R-HSA-997272       pathway  REACTOME  \n",
       "129372          1062       anatomy    UBERON  \n",
       "129373           468       anatomy    UBERON  \n",
       "129374           992       anatomy    UBERON  \n",
       "\n",
       "[129375 rows x 5 columns]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pkg1 = primeKG[['x_index', 'x_name', 'x_id', 'x_type', 'x_source']]\n",
    "pkg2 = primeKG[['y_index', 'y_name', 'y_id', 'y_type', 'y_source']]\n",
    "pkg2.columns = ['x_index', 'x_name', 'x_id', 'x_type', 'x_source']\n",
    "# concat pkg1 and pkg2\n",
    "pkg = pd.concat([pkg1, pkg2], axis=0)\n",
    "pkg = pkg.drop_duplicates()\n",
    "pkg.reset_index(drop=True, inplace=True)\n",
    "pkg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>drug</td>\n",
       "      <td>DrugBank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>HPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>disease</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>disease</td>\n",
       "      <td>MONDO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>biological_process</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>molecular_function</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>exposure</td>\n",
       "      <td>CTD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>pathway</td>\n",
       "      <td>REACTOME</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>anatomy</td>\n",
       "      <td>UBERON</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                x_type       x_source\n",
       "0         gene/protein           NCBI\n",
       "1                 drug       DrugBank\n",
       "2     effect/phenotype            HPO\n",
       "3              disease  MONDO_grouped\n",
       "4              disease          MONDO\n",
       "5   biological_process             GO\n",
       "6   molecular_function             GO\n",
       "7   cellular_component             GO\n",
       "8             exposure            CTD\n",
       "9              pathway       REACTOME\n",
       "10             anatomy         UBERON"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df3 = pkg[['x_type','x_source']]\n",
    "df3 = df3.drop_duplicates()\n",
    "df3.reset_index(drop=True, inplace=True)\n",
    "df3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "pkg.to_csv('data/pkg.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "SDoHKG = pd.read_csv('data/SDoHKG.csv', low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relation</th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_source</th>\n",
       "      <th>y_index</th>\n",
       "      <th>y_name</th>\n",
       "      <th>y_id</th>\n",
       "      <th>y_type</th>\n",
       "      <th>y_source</th>\n",
       "      <th>pmid</th>\n",
       "      <th>probability</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>associates</td>\n",
       "      <td>-1</td>\n",
       "      <td>dementia</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>-1</td>\n",
       "      <td>diagnosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255.0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>associates</td>\n",
       "      <td>-1</td>\n",
       "      <td>dementia</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>-1</td>\n",
       "      <td>disease-modifying treatments</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255.0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>associates</td>\n",
       "      <td>-1</td>\n",
       "      <td>dementia</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>-1</td>\n",
       "      <td>elderly care</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255.0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>associates</td>\n",
       "      <td>-1</td>\n",
       "      <td>dementia</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>-1</td>\n",
       "      <td>guidelines</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255.0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>associates</td>\n",
       "      <td>-1</td>\n",
       "      <td>dementia</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>-1</td>\n",
       "      <td>protocols</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255.0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4053</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>-1</td>\n",
       "      <td>Poverty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>-1</td>\n",
       "      <td>economic stability</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4054</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>-1</td>\n",
       "      <td>Access to Foods That Support Healthy Dietary P...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>-1</td>\n",
       "      <td>neighborhood and built environment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4055</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>-1</td>\n",
       "      <td>Crime and Violence</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>-1</td>\n",
       "      <td>neighborhood and built environment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4056</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>-1</td>\n",
       "      <td>Environmental Conditions</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>-1</td>\n",
       "      <td>neighborhood and built environment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4057</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>-1</td>\n",
       "      <td>Quality of Housing</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>-1</td>\n",
       "      <td>neighborhood and built environment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4058 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         relation  x_index                                             x_name  \\\n",
       "0      associates       -1                                           dementia   \n",
       "1      associates       -1                                           dementia   \n",
       "2      associates       -1                                           dementia   \n",
       "3      associates       -1                                           dementia   \n",
       "4      associates       -1                                           dementia   \n",
       "...           ...      ...                                                ...   \n",
       "4053  subordinate       -1                                            Poverty   \n",
       "4054  subordinate       -1  Access to Foods That Support Healthy Dietary P...   \n",
       "4055  subordinate       -1                                 Crime and Violence   \n",
       "4056  subordinate       -1                           Environmental Conditions   \n",
       "4057  subordinate       -1                                 Quality of Housing   \n",
       "\n",
       "              x_id        x_type       x_source  y_index  \\\n",
       "0     MESH:D003704       Disease      ncbi_mesh       -1   \n",
       "1     MESH:D003704       Disease      ncbi_mesh       -1   \n",
       "2     MESH:D003704       Disease      ncbi_mesh       -1   \n",
       "3     MESH:D003704       Disease      ncbi_mesh       -1   \n",
       "4     MESH:D003704       Disease      ncbi_mesh       -1   \n",
       "...            ...           ...            ...      ...   \n",
       "4053           NaN  SDoH subtype  healthypeople       -1   \n",
       "4054           NaN  SDoH subtype  healthypeople       -1   \n",
       "4055           NaN  SDoH subtype  healthypeople       -1   \n",
       "4056           NaN  SDoH subtype  healthypeople       -1   \n",
       "4057           NaN  SDoH subtype  healthypeople       -1   \n",
       "\n",
       "                                  y_name  y_id       y_type       y_source  \\\n",
       "0                              diagnosis   NaN  SDoH factor  healthypeople   \n",
       "1           disease-modifying treatments   NaN  SDoH factor  healthypeople   \n",
       "2                           elderly care   NaN  SDoH factor  healthypeople   \n",
       "3                             guidelines   NaN  SDoH factor  healthypeople   \n",
       "4                              protocols   NaN  SDoH factor  healthypeople   \n",
       "...                                  ...   ...          ...            ...   \n",
       "4053                  economic stability   NaN    SDoH type  healthypeople   \n",
       "4054  neighborhood and built environment   NaN    SDoH type  healthypeople   \n",
       "4055  neighborhood and built environment   NaN    SDoH type  healthypeople   \n",
       "4056  neighborhood and built environment   NaN    SDoH type  healthypeople   \n",
       "4057  neighborhood and built environment   NaN    SDoH type  healthypeople   \n",
       "\n",
       "            pmid  probability  \n",
       "0     35301255.0          0.5  \n",
       "1     35301255.0          0.5  \n",
       "2     35301255.0          0.5  \n",
       "3     35301255.0          0.5  \n",
       "4     35301255.0          0.5  \n",
       "...          ...          ...  \n",
       "4053         NaN          1.0  \n",
       "4054         NaN          1.0  \n",
       "4055         NaN          1.0  \n",
       "4056         NaN          1.0  \n",
       "4057         NaN          1.0  \n",
       "\n",
       "[4058 rows x 13 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SDoHKG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>id</th>\n",
       "      <th>type</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>dementia</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>patients</td>\n",
       "      <td>9606</td>\n",
       "      <td>Species</td>\n",
       "      <td>ncbi_taxonomy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Alzheimer's disease</td>\n",
       "      <td>MESH:D000544</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>amyloid accumulation</td>\n",
       "      <td>MESH:C000718787</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cerebral vulnerability</td>\n",
       "      <td>MESH:D002547</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>721</th>\n",
       "      <td>hepatitis C</td>\n",
       "      <td>MESH:D019698</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>722</th>\n",
       "      <td>domestic violence/abuse</td>\n",
       "      <td>MESH:D019966</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>723</th>\n",
       "      <td>knee pain</td>\n",
       "      <td>MESH:D046788</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>724</th>\n",
       "      <td>food insecurity</td>\n",
       "      <td>MESH:D005517</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>725</th>\n",
       "      <td>respiratory diseases</td>\n",
       "      <td>MESH:D012140</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>726 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        name               id     type         source\n",
       "0                   dementia     MESH:D003704  Disease      ncbi_mesh\n",
       "1                   patients             9606  Species  ncbi_taxonomy\n",
       "2        Alzheimer's disease     MESH:D000544  Disease      ncbi_mesh\n",
       "3       amyloid accumulation  MESH:C000718787  Disease      ncbi_mesh\n",
       "4     cerebral vulnerability     MESH:D002547  Disease      ncbi_mesh\n",
       "..                       ...              ...      ...            ...\n",
       "721              hepatitis C     MESH:D019698  Disease      ncbi_mesh\n",
       "722  domestic violence/abuse     MESH:D019966  Disease      ncbi_mesh\n",
       "723                knee pain     MESH:D046788  Disease      ncbi_mesh\n",
       "724          food insecurity     MESH:D005517  Disease      ncbi_mesh\n",
       "725     respiratory diseases     MESH:D012140  Disease      ncbi_mesh\n",
       "\n",
       "[726 rows x 4 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get a dataframe that contains all the x_type which are not 'SDoH factor' or 'SDoH subtype'\n",
    "unmatchbio = SDoHKG[~SDoHKG['x_type'].isin(['SDoH factor', 'SDoH subtype'])] \n",
    "unmatchbio = unmatchbio[['x_name', 'x_id', 'x_type', 'x_source']]\n",
    "unmatchbio.columns = ['name', 'id', 'type', 'source']\n",
    "unmatchbio = unmatchbio.drop_duplicates()\n",
    "unmatchbio.reset_index(drop=True, inplace=True)\n",
    "unmatchbio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Species</td>\n",
       "      <td>ncbi_taxonomy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chemical</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Gene</td>\n",
       "      <td>ncbi_gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Disease</td>\n",
       "      <td>omim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Variant</td>\n",
       "      <td>litvar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CellLine</td>\n",
       "      <td>cvcl</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       type         source\n",
       "0   Disease      ncbi_mesh\n",
       "1   Species  ncbi_taxonomy\n",
       "2  Chemical      ncbi_mesh\n",
       "3      Gene      ncbi_gene\n",
       "4   Disease           omim\n",
       "5   Variant         litvar\n",
       "6  CellLine           cvcl"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df4 = unmatchbio[['type','source']]\n",
    "df4 = df4.drop_duplicates()\n",
    "df4.reset_index(drop=True, inplace=True)\n",
    "df4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "unmatchbio.to_csv('data/unmatchbio.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cls</th>\n",
       "      <th>xref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MONDO:0018874</td>\n",
       "      <td>MESH:D015470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MONDO:0008339</td>\n",
       "      <td>MESH:C566738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MONDO:0009233</td>\n",
       "      <td>MESH:C537226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MONDO:0007051</td>\n",
       "      <td>MESH:C535655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MONDO:0011190</td>\n",
       "      <td>MESH:C566582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138714</th>\n",
       "      <td>MONDO:0000978</td>\n",
       "      <td>NCIT:C5854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138715</th>\n",
       "      <td>MONDO:0016646</td>\n",
       "      <td>UMLS:C4273829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138716</th>\n",
       "      <td>MONDO:0016646</td>\n",
       "      <td>MEDGEN:895207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138717</th>\n",
       "      <td>MONDO:0016646</td>\n",
       "      <td>ORPHANET:250932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138718</th>\n",
       "      <td>MONDO:0016646</td>\n",
       "      <td>GARD:20686</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>130944 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  cls             xref\n",
       "0       MONDO:0018874     MESH:D015470\n",
       "1       MONDO:0008339     MESH:C566738\n",
       "2       MONDO:0009233     MESH:C537226\n",
       "3       MONDO:0007051     MESH:C535655\n",
       "4       MONDO:0011190     MESH:C566582\n",
       "...               ...              ...\n",
       "138714  MONDO:0000978       NCIT:C5854\n",
       "138715  MONDO:0016646    UMLS:C4273829\n",
       "138716  MONDO:0016646    MEDGEN:895207\n",
       "138717  MONDO:0016646  ORPHANET:250932\n",
       "138718  MONDO:0016646       GARD:20686\n",
       "\n",
       "[130944 rows x 2 columns]"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xrefs = pd.read_csv('data/xrefs.tsv', sep='\\t')\n",
    "meshMONDO = pd.read_csv('data/meshMONDO.csv', low_memory=False)\n",
    "meshMONDO = meshMONDO[['mapped_curie', 'curie_id']]\n",
    "meshMONDO.columns = ['cls', 'xref']\n",
    "meshMONDO = pd.concat([meshMONDO, xrefs], axis=0).reset_index(drop=True)\n",
    "# upper case the 'xref' column\n",
    "meshMONDO['xref'] = meshMONDO['xref'].str.upper()\n",
    "meshMONDO = meshMONDO.drop_duplicates()\n",
    "meshMONDO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "meshMONDO.to_csv('data/MESH-MONDO.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "meshMONDO = pd.read_csv('data/MESH-MONDO.csv', low_memory=False)\n",
    "meshDB = pd.read_csv('data/Mesh2DrugBank_bioportal.csv', low_memory=False)\n",
    "pkg = pd.read_csv('data/pkg.csv')\n",
    "unmatchbio = pd.read_csv('data/unmatchbio.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each row in unmatchbio, find the correspongding row in pkg\n",
    "\n",
    "gene: direct match\n",
    "\n",
    "mesh&disease: use MESH-MONDO.cv, mondo/grouped(some mesh id don't have a matched mondo id), https://github.com/natacourby/Disease_ontologies_for_knowledge_graphs\n",
    "\n",
    "omim&disease:\n",
    "\n",
    "mesh&chemical: drugbank\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>id</th>\n",
       "      <th>type</th>\n",
       "      <th>source</th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>dementia</td>\n",
       "      <td>MESH:D003704</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>37573</td>\n",
       "      <td>dementia (disease)</td>\n",
       "      <td>1627</td>\n",
       "      <td>disease</td>\n",
       "      <td>MONDO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>patients</td>\n",
       "      <td>9606</td>\n",
       "      <td>Species</td>\n",
       "      <td>ncbi_taxonomy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>patients</td>\n",
       "      <td>9606</td>\n",
       "      <td>Species</td>\n",
       "      <td>ncbi_taxonomy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Alzheimer's disease</td>\n",
       "      <td>MESH:D000544</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>28780</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>11913_7088_11743_12153_7089_11194_12321_15140_...</td>\n",
       "      <td>disease</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>amyloid accumulation</td>\n",
       "      <td>MESH:C000718787</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>NaN</td>\n",
       "      <td>amyloid accumulation</td>\n",
       "      <td>MESH:C000718787</td>\n",
       "      <td>disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cerebral vulnerability</td>\n",
       "      <td>MESH:D002547</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>33667</td>\n",
       "      <td>cerebral palsy</td>\n",
       "      <td>6497</td>\n",
       "      <td>disease</td>\n",
       "      <td>MONDO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>721</th>\n",
       "      <td>hepatitis C</td>\n",
       "      <td>MESH:D019698</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>83799</td>\n",
       "      <td>chronic hepatitis C virus infection</td>\n",
       "      <td>5354</td>\n",
       "      <td>disease</td>\n",
       "      <td>MONDO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>722</th>\n",
       "      <td>domestic violence/abuse</td>\n",
       "      <td>MESH:D019966</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>36693</td>\n",
       "      <td>substance-related disorder</td>\n",
       "      <td>2494</td>\n",
       "      <td>disease</td>\n",
       "      <td>MONDO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>723</th>\n",
       "      <td>knee pain</td>\n",
       "      <td>MESH:D046788</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>97538</td>\n",
       "      <td>patellofemoral pain syndrome</td>\n",
       "      <td>6894</td>\n",
       "      <td>disease</td>\n",
       "      <td>MONDO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>724</th>\n",
       "      <td>food insecurity</td>\n",
       "      <td>MESH:D005517</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>NaN</td>\n",
       "      <td>food insecurity</td>\n",
       "      <td>MESH:D005517</td>\n",
       "      <td>disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>725</th>\n",
       "      <td>respiratory diseases</td>\n",
       "      <td>MESH:D012140</td>\n",
       "      <td>Disease</td>\n",
       "      <td>ncbi_mesh</td>\n",
       "      <td>35553</td>\n",
       "      <td>respiratory system disease</td>\n",
       "      <td>5087</td>\n",
       "      <td>disease</td>\n",
       "      <td>MONDO</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>726 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        name               id     type         source x_index  \\\n",
       "0                   dementia     MESH:D003704  Disease      ncbi_mesh   37573   \n",
       "1                   patients             9606  Species  ncbi_taxonomy     NaN   \n",
       "2        Alzheimer's disease     MESH:D000544  Disease      ncbi_mesh   28780   \n",
       "3       amyloid accumulation  MESH:C000718787  Disease      ncbi_mesh     NaN   \n",
       "4     cerebral vulnerability     MESH:D002547  Disease      ncbi_mesh   33667   \n",
       "..                       ...              ...      ...            ...     ...   \n",
       "721              hepatitis C     MESH:D019698  Disease      ncbi_mesh   83799   \n",
       "722  domestic violence/abuse     MESH:D019966  Disease      ncbi_mesh   36693   \n",
       "723                knee pain     MESH:D046788  Disease      ncbi_mesh   97538   \n",
       "724          food insecurity     MESH:D005517  Disease      ncbi_mesh     NaN   \n",
       "725     respiratory diseases     MESH:D012140  Disease      ncbi_mesh   35553   \n",
       "\n",
       "                                  x_name  \\\n",
       "0                     dementia (disease)   \n",
       "1                               patients   \n",
       "2                      Alzheimer disease   \n",
       "3                   amyloid accumulation   \n",
       "4                         cerebral palsy   \n",
       "..                                   ...   \n",
       "721  chronic hepatitis C virus infection   \n",
       "722           substance-related disorder   \n",
       "723         patellofemoral pain syndrome   \n",
       "724                      food insecurity   \n",
       "725           respiratory system disease   \n",
       "\n",
       "                                                  x_id   x_type       x_source  \n",
       "0                                                 1627  disease          MONDO  \n",
       "1                                                 9606  Species  ncbi_taxonomy  \n",
       "2    11913_7088_11743_12153_7089_11194_12321_15140_...  disease  MONDO_grouped  \n",
       "3                                      MESH:C000718787  disease      ncbi_mesh  \n",
       "4                                                 6497  disease          MONDO  \n",
       "..                                                 ...      ...            ...  \n",
       "721                                               5354  disease          MONDO  \n",
       "722                                               2494  disease          MONDO  \n",
       "723                                               6894  disease          MONDO  \n",
       "724                                       MESH:D005517  disease      ncbi_mesh  \n",
       "725                                               5087  disease          MONDO  \n",
       "\n",
       "[726 rows x 9 columns]"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import re\n",
    "updated_rows = []\n",
    "for index, row in unmatchbio.iterrows():\n",
    "    if row['source'] == 'ncbi_gene':\n",
    "        match = pkg[(pkg['x_id'] == row['id']) & (pkg['x_source'] == 'NCBI')]\n",
    "        if not match.empty:\n",
    "            # Add the target DataFrame's columns to the row\n",
    "            for col in match.columns:\n",
    "                row[col] = match[col].values[0]\n",
    "        else: \n",
    "            row['x_index'] = 'NaN'\n",
    "            row['x_name'] = row['name']\n",
    "            row['x_id'] = row['id']\n",
    "            row['x_type'] = 'gene/protein'\n",
    "            row['x_source'] = 'NCBI'\n",
    "\n",
    "    elif row['source'] == 'ncbi_mesh' and row['type'] == 'Disease': # map mesh disease to MONDO\n",
    "        # Check if the id is in meshMONDO['curie_id']\n",
    "        match = meshMONDO[meshMONDO['xref'].str.lower() == row['id'].lower()]\n",
    "        if not match.empty:\n",
    "            # Get the corresponding meshMONDO['mapped_curie']\n",
    "            mapped_curie = match['cls'].values[0]\n",
    "            # Extract the digits from the right until zero\n",
    "            match_digits = re.search(r'0*(\\d+)', mapped_curie.split(':')[-1])\n",
    "            extracted_digits = match_digits.group(1)\n",
    "            # MONDO entities that are not in PrimeKG\n",
    "            row['x_index'] = 'NaN'\n",
    "            row['x_name'] = row['name']\n",
    "            row['x_id'] = extracted_digits\n",
    "            row['x_type'] = 'disease'\n",
    "            row['x_source'] = 'MONDO'\n",
    "            # x_source is MONDO\n",
    "            pkg_match = pkg[(pkg['x_id'].astype(str) == extracted_digits) & (pkg['x_source'] == 'MONDO')]\n",
    "            if not pkg_match.empty:\n",
    "                # Add the target DataFrame's columns to the row\n",
    "                for col in pkg_match.columns:\n",
    "                    row[col] = pkg_match[col].values[0]\n",
    "            # x_source is MONDO_grouped\n",
    "            pkg['x_id_split'] = pkg['x_id'].apply(lambda x: x.split('_'))\n",
    "            pkg_match1 = pkg[(pkg['x_id_split'].apply(lambda x: extracted_digits in x)) & (pkg['x_source'] == 'MONDO_grouped')]\n",
    "            if not pkg_match1.empty:\n",
    "                # Add the target DataFrame's columns to the row\n",
    "                for col in pkg_match1.columns:\n",
    "                    row[col] = pkg_match1[col].values[0]\n",
    "            # Drop the temporary 'x_id_split' column\n",
    "            pkg.drop(columns=['x_id_split'], inplace=True)\n",
    "        else:\n",
    "            row['x_index'] = 'NaN'\n",
    "            row['x_name'] = row['name']\n",
    "            row['x_id'] = row['id']\n",
    "            row['x_type'] = 'disease'\n",
    "            row['x_source'] = 'ncbi_mesh'\n",
    "    elif row['source'] == 'omim' and row['type'] == 'Disease': # map omim disease to MONDO\n",
    "        pass\n",
    "    elif row['source'] == 'ncbi_mesh' and row['type'] == 'Chemical': # map mesh chemical to drugbank\n",
    "        id = row['id'].split(':')[1]\n",
    "        match = meshDB[meshDB['Mesh_ID'] == id]\n",
    "        if not match.empty:\n",
    "            mapped_curie = match['drugBank_ID'].values[0]\n",
    "            # DB entities that are not in PrimeKG\n",
    "            row['x_index'] = 'NaN'\n",
    "            row['x_name'] = row['name']\n",
    "            row['x_id'] = mapped_curie\n",
    "            row['x_type'] = 'drug'\n",
    "            row['x_source'] = 'DrugBank'\n",
    "            pkg_match = pkg[(pkg['x_id'].astype(str) == mapped_curie) & (pkg['x_source'] == 'DrugBank')]\n",
    "            if not pkg_match.empty:\n",
    "                # Add the target DataFrame's columns to the row\n",
    "                for col in pkg_match.columns:\n",
    "                    row[col] = pkg_match[col].values[0]\n",
    "        else:\n",
    "            row['x_index'] = 'NaN'\n",
    "            row['x_name'] = row['name']\n",
    "            row['x_id'] = row['id']\n",
    "            row['x_type'] = 'drug'\n",
    "            row['x_source'] = 'ncbi_mesh'\n",
    "\n",
    "    # match for other types that are not in PrimeKG\n",
    "    else: \n",
    "        row['x_index'] = 'NaN'\n",
    "        row['x_name'] = row['name']\n",
    "        row['x_id'] = row['id']\n",
    "        row['x_type'] = row['type']\n",
    "        row['x_source'] = row['source']\n",
    "\n",
    "    updated_rows.append(row)\n",
    "\n",
    "updated_df = pd.DataFrame(updated_rows)\n",
    "# drop columns 'x_id'\n",
    "updated_df.drop(columns=['x_id_split'], inplace=True)\n",
    "updated_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "updated_df.to_csv('data/updated_unmatchbio.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "updated_df = pd.read_csv('data/matchbio.csv', low_memory=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "replace the corresponding information in SDoHKG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relation</th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_source</th>\n",
       "      <th>y_index</th>\n",
       "      <th>y_name</th>\n",
       "      <th>y_id</th>\n",
       "      <th>y_type</th>\n",
       "      <th>y_source</th>\n",
       "      <th>pmid</th>\n",
       "      <th>probability</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>associates</td>\n",
       "      <td>37573.0</td>\n",
       "      <td>dementia (disease)</td>\n",
       "      <td>1627</td>\n",
       "      <td>disease</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>-1</td>\n",
       "      <td>diagnosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255.0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>associates</td>\n",
       "      <td>37573.0</td>\n",
       "      <td>dementia (disease)</td>\n",
       "      <td>1627</td>\n",
       "      <td>disease</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>-1</td>\n",
       "      <td>disease-modifying treatments</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255.0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>associates</td>\n",
       "      <td>37573.0</td>\n",
       "      <td>dementia (disease)</td>\n",
       "      <td>1627</td>\n",
       "      <td>disease</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>-1</td>\n",
       "      <td>elderly care</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255.0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>associates</td>\n",
       "      <td>37573.0</td>\n",
       "      <td>dementia (disease)</td>\n",
       "      <td>1627</td>\n",
       "      <td>disease</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>-1</td>\n",
       "      <td>guidelines</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255.0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>associates</td>\n",
       "      <td>37573.0</td>\n",
       "      <td>dementia (disease)</td>\n",
       "      <td>1627</td>\n",
       "      <td>disease</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>-1</td>\n",
       "      <td>protocols</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>35301255.0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4053</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>Poverty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>-1</td>\n",
       "      <td>economic stability</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4054</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>Access to Foods That Support Healthy Dietary P...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>-1</td>\n",
       "      <td>neighborhood and built environment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4055</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>Crime and Violence</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>-1</td>\n",
       "      <td>neighborhood and built environment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4056</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>Environmental Conditions</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>-1</td>\n",
       "      <td>neighborhood and built environment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4057</th>\n",
       "      <td>subordinate</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>Quality of Housing</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>-1</td>\n",
       "      <td>neighborhood and built environment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4058 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         relation  x_index                                             x_name  \\\n",
       "0      associates  37573.0                                 dementia (disease)   \n",
       "1      associates  37573.0                                 dementia (disease)   \n",
       "2      associates  37573.0                                 dementia (disease)   \n",
       "3      associates  37573.0                                 dementia (disease)   \n",
       "4      associates  37573.0                                 dementia (disease)   \n",
       "...           ...      ...                                                ...   \n",
       "4053  subordinate     -1.0                                            Poverty   \n",
       "4054  subordinate     -1.0  Access to Foods That Support Healthy Dietary P...   \n",
       "4055  subordinate     -1.0                                 Crime and Violence   \n",
       "4056  subordinate     -1.0                           Environmental Conditions   \n",
       "4057  subordinate     -1.0                                 Quality of Housing   \n",
       "\n",
       "      x_id        x_type       x_source  y_index  \\\n",
       "0     1627       disease          MONDO       -1   \n",
       "1     1627       disease          MONDO       -1   \n",
       "2     1627       disease          MONDO       -1   \n",
       "3     1627       disease          MONDO       -1   \n",
       "4     1627       disease          MONDO       -1   \n",
       "...    ...           ...            ...      ...   \n",
       "4053   NaN  SDoH subtype  healthypeople       -1   \n",
       "4054   NaN  SDoH subtype  healthypeople       -1   \n",
       "4055   NaN  SDoH subtype  healthypeople       -1   \n",
       "4056   NaN  SDoH subtype  healthypeople       -1   \n",
       "4057   NaN  SDoH subtype  healthypeople       -1   \n",
       "\n",
       "                                  y_name  y_id       y_type       y_source  \\\n",
       "0                              diagnosis   NaN  SDoH factor  healthypeople   \n",
       "1           disease-modifying treatments   NaN  SDoH factor  healthypeople   \n",
       "2                           elderly care   NaN  SDoH factor  healthypeople   \n",
       "3                             guidelines   NaN  SDoH factor  healthypeople   \n",
       "4                              protocols   NaN  SDoH factor  healthypeople   \n",
       "...                                  ...   ...          ...            ...   \n",
       "4053                  economic stability   NaN    SDoH type  healthypeople   \n",
       "4054  neighborhood and built environment   NaN    SDoH type  healthypeople   \n",
       "4055  neighborhood and built environment   NaN    SDoH type  healthypeople   \n",
       "4056  neighborhood and built environment   NaN    SDoH type  healthypeople   \n",
       "4057  neighborhood and built environment   NaN    SDoH type  healthypeople   \n",
       "\n",
       "            pmid  probability  \n",
       "0     35301255.0          0.5  \n",
       "1     35301255.0          0.5  \n",
       "2     35301255.0          0.5  \n",
       "3     35301255.0          0.5  \n",
       "4     35301255.0          0.5  \n",
       "...          ...          ...  \n",
       "4053         NaN          1.0  \n",
       "4054         NaN          1.0  \n",
       "4055         NaN          1.0  \n",
       "4056         NaN          1.0  \n",
       "4057         NaN          1.0  \n",
       "\n",
       "[4058 rows x 13 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for index, row in SDoHKG.iterrows(): \n",
    "    updated_row = updated_df[(updated_df['name'] == row['x_name']) & (updated_df['id'] == row['x_id'])]\n",
    "    if not updated_row.empty:\n",
    "        updated_row = updated_row.iloc[0]\n",
    "        SDoHKG.at[index, 'x_index'] = updated_row['x_index']\n",
    "        SDoHKG.at[index, 'x_name'] = updated_row['x_name']\n",
    "        SDoHKG.at[index, 'x_id'] = updated_row['x_id']\n",
    "        SDoHKG.at[index, 'x_type'] = updated_row['x_type']\n",
    "        SDoHKG.at[index, 'x_source'] = updated_row['x_source']\n",
    "        \n",
    "SDoHKG.to_csv('data/SDoHKG_matched.csv', index=False)\n",
    "SDoHKG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find whether there are nan in the 'x_name' column\n",
    "SDoHKG['y_source'].isnull().values.any()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### get primeKG subgraph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relation</th>\n",
       "      <th>display_relation</th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_source</th>\n",
       "      <th>y_index</th>\n",
       "      <th>y_id</th>\n",
       "      <th>y_type</th>\n",
       "      <th>y_name</th>\n",
       "      <th>y_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>379115</th>\n",
       "      <td>contraindication</td>\n",
       "      <td>contraindication</td>\n",
       "      <td>20294</td>\n",
       "      <td>DB11160</td>\n",
       "      <td>drug</td>\n",
       "      <td>Phenyltoloxamine</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>28780</td>\n",
       "      <td>11913_7088_11743_12153_7089_11194_12321_15140_...</td>\n",
       "      <td>disease</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379119</th>\n",
       "      <td>contraindication</td>\n",
       "      <td>contraindication</td>\n",
       "      <td>16490</td>\n",
       "      <td>DB01620</td>\n",
       "      <td>drug</td>\n",
       "      <td>Pheniramine</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>28780</td>\n",
       "      <td>11913_7088_11743_12153_7089_11194_12321_15140_...</td>\n",
       "      <td>disease</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379123</th>\n",
       "      <td>contraindication</td>\n",
       "      <td>contraindication</td>\n",
       "      <td>15147</td>\n",
       "      <td>DB01173</td>\n",
       "      <td>drug</td>\n",
       "      <td>Orphenadrine</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>28780</td>\n",
       "      <td>11913_7088_11743_12153_7089_11194_12321_15140_...</td>\n",
       "      <td>disease</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379127</th>\n",
       "      <td>contraindication</td>\n",
       "      <td>contraindication</td>\n",
       "      <td>17407</td>\n",
       "      <td>DB00175</td>\n",
       "      <td>drug</td>\n",
       "      <td>Pravastatin</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>28780</td>\n",
       "      <td>11913_7088_11743_12153_7089_11194_12321_15140_...</td>\n",
       "      <td>disease</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379131</th>\n",
       "      <td>contraindication</td>\n",
       "      <td>contraindication</td>\n",
       "      <td>15832</td>\n",
       "      <td>DB00395</td>\n",
       "      <td>drug</td>\n",
       "      <td>Carisoprodol</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>28780</td>\n",
       "      <td>11913_7088_11743_12153_7089_11194_12321_15140_...</td>\n",
       "      <td>disease</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5921563</th>\n",
       "      <td>disease_phenotype_positive</td>\n",
       "      <td>phenotype present</td>\n",
       "      <td>22671</td>\n",
       "      <td>11446</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Abnormality of higher mental function</td>\n",
       "      <td>HPO</td>\n",
       "      <td>28780</td>\n",
       "      <td>11913_7088_11743_12153_7089_11194_12321_15140_...</td>\n",
       "      <td>disease</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5921564</th>\n",
       "      <td>disease_phenotype_positive</td>\n",
       "      <td>phenotype present</td>\n",
       "      <td>90327</td>\n",
       "      <td>30219</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>Semantic dementia</td>\n",
       "      <td>HPO</td>\n",
       "      <td>28780</td>\n",
       "      <td>11913_7088_11743_12153_7089_11194_12321_15140_...</td>\n",
       "      <td>disease</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6046106</th>\n",
       "      <td>disease_disease</td>\n",
       "      <td>parent-child</td>\n",
       "      <td>94982</td>\n",
       "      <td>10783</td>\n",
       "      <td>disease</td>\n",
       "      <td>Alzheimer disease, susceptibility to, mitochon...</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>28780</td>\n",
       "      <td>11913_7088_11743_12153_7089_11194_12321_15140_...</td>\n",
       "      <td>disease</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6046901</th>\n",
       "      <td>disease_disease</td>\n",
       "      <td>parent-child</td>\n",
       "      <td>83960</td>\n",
       "      <td>11401</td>\n",
       "      <td>disease</td>\n",
       "      <td>Alzheimer disease without neurofibrillary tangles</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>28780</td>\n",
       "      <td>11913_7088_11743_12153_7089_11194_12321_15140_...</td>\n",
       "      <td>disease</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6047037</th>\n",
       "      <td>disease_disease</td>\n",
       "      <td>parent-child</td>\n",
       "      <td>31527</td>\n",
       "      <td>11513</td>\n",
       "      <td>disease</td>\n",
       "      <td>Alzheimer disease, familial early-onset, with ...</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>28780</td>\n",
       "      <td>11913_7088_11743_12153_7089_11194_12321_15140_...</td>\n",
       "      <td>disease</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>233 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           relation   display_relation  x_index     x_id  \\\n",
       "379115             contraindication   contraindication    20294  DB11160   \n",
       "379119             contraindication   contraindication    16490  DB01620   \n",
       "379123             contraindication   contraindication    15147  DB01173   \n",
       "379127             contraindication   contraindication    17407  DB00175   \n",
       "379131             contraindication   contraindication    15832  DB00395   \n",
       "...                             ...                ...      ...      ...   \n",
       "5921563  disease_phenotype_positive  phenotype present    22671    11446   \n",
       "5921564  disease_phenotype_positive  phenotype present    90327    30219   \n",
       "6046106             disease_disease       parent-child    94982    10783   \n",
       "6046901             disease_disease       parent-child    83960    11401   \n",
       "6047037             disease_disease       parent-child    31527    11513   \n",
       "\n",
       "                   x_type                                             x_name  \\\n",
       "379115               drug                                   Phenyltoloxamine   \n",
       "379119               drug                                        Pheniramine   \n",
       "379123               drug                                       Orphenadrine   \n",
       "379127               drug                                        Pravastatin   \n",
       "379131               drug                                       Carisoprodol   \n",
       "...                   ...                                                ...   \n",
       "5921563  effect/phenotype              Abnormality of higher mental function   \n",
       "5921564  effect/phenotype                                  Semantic dementia   \n",
       "6046106           disease  Alzheimer disease, susceptibility to, mitochon...   \n",
       "6046901           disease  Alzheimer disease without neurofibrillary tangles   \n",
       "6047037           disease  Alzheimer disease, familial early-onset, with ...   \n",
       "\n",
       "         x_source  y_index                                               y_id  \\\n",
       "379115   DrugBank    28780  11913_7088_11743_12153_7089_11194_12321_15140_...   \n",
       "379119   DrugBank    28780  11913_7088_11743_12153_7089_11194_12321_15140_...   \n",
       "379123   DrugBank    28780  11913_7088_11743_12153_7089_11194_12321_15140_...   \n",
       "379127   DrugBank    28780  11913_7088_11743_12153_7089_11194_12321_15140_...   \n",
       "379131   DrugBank    28780  11913_7088_11743_12153_7089_11194_12321_15140_...   \n",
       "...           ...      ...                                                ...   \n",
       "5921563       HPO    28780  11913_7088_11743_12153_7089_11194_12321_15140_...   \n",
       "5921564       HPO    28780  11913_7088_11743_12153_7089_11194_12321_15140_...   \n",
       "6046106     MONDO    28780  11913_7088_11743_12153_7089_11194_12321_15140_...   \n",
       "6046901     MONDO    28780  11913_7088_11743_12153_7089_11194_12321_15140_...   \n",
       "6047037     MONDO    28780  11913_7088_11743_12153_7089_11194_12321_15140_...   \n",
       "\n",
       "          y_type             y_name       y_source  \n",
       "379115   disease  Alzheimer disease  MONDO_grouped  \n",
       "379119   disease  Alzheimer disease  MONDO_grouped  \n",
       "379123   disease  Alzheimer disease  MONDO_grouped  \n",
       "379127   disease  Alzheimer disease  MONDO_grouped  \n",
       "379131   disease  Alzheimer disease  MONDO_grouped  \n",
       "...          ...                ...            ...  \n",
       "5921563  disease  Alzheimer disease  MONDO_grouped  \n",
       "5921564  disease  Alzheimer disease  MONDO_grouped  \n",
       "6046106  disease  Alzheimer disease  MONDO_grouped  \n",
       "6046901  disease  Alzheimer disease  MONDO_grouped  \n",
       "6047037  disease  Alzheimer disease  MONDO_grouped  \n",
       "\n",
       "[233 rows x 12 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "primekg = pd.read_csv('data\\kg.csv', low_memory=False)\n",
    "alzheimer = primekg.query('y_name==\"Alzheimer disease\"|x_name==\"Alzheime disease\"')\n",
    "alzheimer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_hops(df, start_node, max_hops=6):\n",
    "    visited = set()  # Set to track visited nodes\n",
    "    current_nodes = set([start_node])  # Start with the initial node\n",
    "\n",
    "    for _ in range(max_hops):\n",
    "        new_nodes = set()\n",
    "        # Find all rows where the current set of nodes are mentioned in x_name or y_name\n",
    "        query = df[(df['x_name'].isin(current_nodes)) | (df['y_name'].isin(current_nodes))]\n",
    "        # Add all connected nodes to the new set\n",
    "        new_nodes.update(query['x_name'].tolist())\n",
    "        new_nodes.update(query['y_name'].tolist())\n",
    "        # Update the visited nodes\n",
    "        visited.update(current_nodes)\n",
    "        # Update current_nodes with new nodes that haven't been visited\n",
    "        current_nodes = new_nodes - visited\n",
    "\n",
    "    return df[(df['x_name'].isin(visited)) | (df['y_name'].isin(visited))]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get 6 hops from 'Alzheimer disease'\n",
    "alzheimer_hops = get_hops(primekg, 'Alzheimer disease', max_hops=4)\n",
    "\n",
    "# Save the sub-graph to a new CSV file\n",
    "alzheimer_hops.to_csv('data/alzheimer_sub_graph4.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relation</th>\n",
       "      <th>display_relation</th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_source</th>\n",
       "      <th>y_index</th>\n",
       "      <th>y_id</th>\n",
       "      <th>y_type</th>\n",
       "      <th>y_name</th>\n",
       "      <th>y_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>29</td>\n",
       "      <td>2147</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>F2</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>844</td>\n",
       "      <td>6696</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>SPP1</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>16</td>\n",
       "      <td>6633</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>SNRPD2</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>216</td>\n",
       "      <td>23385</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCSTN</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>33</td>\n",
       "      <td>2932</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>GSK3B</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>4817</td>\n",
       "      <td>4853</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NOTCH2</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>33</td>\n",
       "      <td>2932</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>GSK3B</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>8918</td>\n",
       "      <td>11033</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>ADAP1</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>61</td>\n",
       "      <td>801</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>CALM1</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>2621</td>\n",
       "      <td>3845</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246617</th>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>expression absent</td>\n",
       "      <td>63180</td>\n",
       "      <td>81</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>metanephros</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>2329</td>\n",
       "      <td>7124</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>TNF</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246618</th>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>expression absent</td>\n",
       "      <td>63824</td>\n",
       "      <td>1377</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>quadriceps femoris</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>2329</td>\n",
       "      <td>7124</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>TNF</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246619</th>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>expression absent</td>\n",
       "      <td>63826</td>\n",
       "      <td>1379</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>vastus lateralis</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>213</td>\n",
       "      <td>3630</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>INS</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246620</th>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>expression absent</td>\n",
       "      <td>63952</td>\n",
       "      <td>1507</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>biceps brachii</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>213</td>\n",
       "      <td>3630</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>INS</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246621</th>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>expression absent</td>\n",
       "      <td>64778</td>\n",
       "      <td>2349</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>myocardium</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>213</td>\n",
       "      <td>3630</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>INS</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>246622 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      relation   display_relation  x_index  x_id  \\\n",
       "0              protein_protein                ppi       29  2147   \n",
       "1              protein_protein                ppi       16  6633   \n",
       "2              protein_protein                ppi       33  2932   \n",
       "3              protein_protein                ppi       33  2932   \n",
       "4              protein_protein                ppi       61   801   \n",
       "...                        ...                ...      ...   ...   \n",
       "246617  anatomy_protein_absent  expression absent    63180    81   \n",
       "246618  anatomy_protein_absent  expression absent    63824  1377   \n",
       "246619  anatomy_protein_absent  expression absent    63826  1379   \n",
       "246620  anatomy_protein_absent  expression absent    63952  1507   \n",
       "246621  anatomy_protein_absent  expression absent    64778  2349   \n",
       "\n",
       "              x_type              x_name x_source  y_index   y_id  \\\n",
       "0       gene/protein                  F2     NCBI      844   6696   \n",
       "1       gene/protein              SNRPD2     NCBI      216  23385   \n",
       "2       gene/protein               GSK3B     NCBI     4817   4853   \n",
       "3       gene/protein               GSK3B     NCBI     8918  11033   \n",
       "4       gene/protein               CALM1     NCBI     2621   3845   \n",
       "...              ...                 ...      ...      ...    ...   \n",
       "246617       anatomy         metanephros   UBERON     2329   7124   \n",
       "246618       anatomy  quadriceps femoris   UBERON     2329   7124   \n",
       "246619       anatomy    vastus lateralis   UBERON      213   3630   \n",
       "246620       anatomy      biceps brachii   UBERON      213   3630   \n",
       "246621       anatomy          myocardium   UBERON      213   3630   \n",
       "\n",
       "              y_type  y_name y_source  \n",
       "0       gene/protein    SPP1     NCBI  \n",
       "1       gene/protein   NCSTN     NCBI  \n",
       "2       gene/protein  NOTCH2     NCBI  \n",
       "3       gene/protein   ADAP1     NCBI  \n",
       "4       gene/protein    KRAS     NCBI  \n",
       "...              ...     ...      ...  \n",
       "246617  gene/protein     TNF     NCBI  \n",
       "246618  gene/protein     TNF     NCBI  \n",
       "246619  gene/protein     INS     NCBI  \n",
       "246620  gene/protein     INS     NCBI  \n",
       "246621  gene/protein     INS     NCBI  \n",
       "\n",
       "[246622 rows x 12 columns]"
      ]
     },
     "execution_count": 175,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alzheimer_hops = pd.read_csv('data/alzheimer_sub_graph2.csv', low_memory=False)\n",
    "alzheimer_hops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relation</th>\n",
       "      <th>display_relation</th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_source</th>\n",
       "      <th>y_index</th>\n",
       "      <th>y_id</th>\n",
       "      <th>y_type</th>\n",
       "      <th>y_name</th>\n",
       "      <th>y_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>29</td>\n",
       "      <td>2147</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>F2</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>844</td>\n",
       "      <td>6696</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>SPP1</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>16</td>\n",
       "      <td>6633</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>SNRPD2</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>216</td>\n",
       "      <td>23385</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCSTN</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>33</td>\n",
       "      <td>2932</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>GSK3B</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>4817</td>\n",
       "      <td>4853</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NOTCH2</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>33</td>\n",
       "      <td>2932</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>GSK3B</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>8918</td>\n",
       "      <td>11033</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>ADAP1</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>61</td>\n",
       "      <td>801</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>CALM1</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>2621</td>\n",
       "      <td>3845</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246617</th>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>expression absent</td>\n",
       "      <td>63180</td>\n",
       "      <td>81</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>metanephros</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>2329</td>\n",
       "      <td>7124</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>TNF</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246618</th>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>expression absent</td>\n",
       "      <td>63824</td>\n",
       "      <td>1377</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>quadriceps femoris</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>2329</td>\n",
       "      <td>7124</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>TNF</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246619</th>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>expression absent</td>\n",
       "      <td>63826</td>\n",
       "      <td>1379</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>vastus lateralis</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>213</td>\n",
       "      <td>3630</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>INS</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246620</th>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>expression absent</td>\n",
       "      <td>63952</td>\n",
       "      <td>1507</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>biceps brachii</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>213</td>\n",
       "      <td>3630</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>INS</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246621</th>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>expression absent</td>\n",
       "      <td>64778</td>\n",
       "      <td>2349</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>myocardium</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>213</td>\n",
       "      <td>3630</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>INS</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>245252 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      relation   display_relation  x_index  x_id  \\\n",
       "0              protein_protein                ppi       29  2147   \n",
       "1              protein_protein                ppi       16  6633   \n",
       "2              protein_protein                ppi       33  2932   \n",
       "3              protein_protein                ppi       33  2932   \n",
       "4              protein_protein                ppi       61   801   \n",
       "...                        ...                ...      ...   ...   \n",
       "246617  anatomy_protein_absent  expression absent    63180    81   \n",
       "246618  anatomy_protein_absent  expression absent    63824  1377   \n",
       "246619  anatomy_protein_absent  expression absent    63826  1379   \n",
       "246620  anatomy_protein_absent  expression absent    63952  1507   \n",
       "246621  anatomy_protein_absent  expression absent    64778  2349   \n",
       "\n",
       "              x_type              x_name x_source  y_index   y_id  \\\n",
       "0       gene/protein                  F2     NCBI      844   6696   \n",
       "1       gene/protein              SNRPD2     NCBI      216  23385   \n",
       "2       gene/protein               GSK3B     NCBI     4817   4853   \n",
       "3       gene/protein               GSK3B     NCBI     8918  11033   \n",
       "4       gene/protein               CALM1     NCBI     2621   3845   \n",
       "...              ...                 ...      ...      ...    ...   \n",
       "246617       anatomy         metanephros   UBERON     2329   7124   \n",
       "246618       anatomy  quadriceps femoris   UBERON     2329   7124   \n",
       "246619       anatomy    vastus lateralis   UBERON      213   3630   \n",
       "246620       anatomy      biceps brachii   UBERON      213   3630   \n",
       "246621       anatomy          myocardium   UBERON      213   3630   \n",
       "\n",
       "              y_type  y_name y_source  \n",
       "0       gene/protein    SPP1     NCBI  \n",
       "1       gene/protein   NCSTN     NCBI  \n",
       "2       gene/protein  NOTCH2     NCBI  \n",
       "3       gene/protein   ADAP1     NCBI  \n",
       "4       gene/protein    KRAS     NCBI  \n",
       "...              ...     ...      ...  \n",
       "246617  gene/protein     TNF     NCBI  \n",
       "246618  gene/protein     TNF     NCBI  \n",
       "246619  gene/protein     INS     NCBI  \n",
       "246620  gene/protein     INS     NCBI  \n",
       "246621  gene/protein     INS     NCBI  \n",
       "\n",
       "[245252 rows x 12 columns]"
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get the rows where the x_type or y_type is 'gene/protein','disease','drug'\n",
    "alzheimer_sub = alzheimer_hops[alzheimer_hops['x_type'].isin(['gene/protein', 'disease', 'drug']) | alzheimer_hops['y_type'].isin(['gene/protein', 'disease', 'drug'])]\n",
    "alzheimer_sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {},
   "outputs": [],
   "source": [
    "alzheimer_sub.to_csv('data/alzheimer_sub_graph.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### generate dataset for link prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "SDoHKG = pd.read_csv('data/SDoHKG_matched.csv', low_memory=False)\n",
    "alzheimer_sub = pd.read_csv('data/alzheimer_sub_graph.csv', low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>relation</th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_source</th>\n",
       "      <th>y_index</th>\n",
       "      <th>y_name</th>\n",
       "      <th>y_id</th>\n",
       "      <th>y_type</th>\n",
       "      <th>y_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>protein_protein</td>\n",
       "      <td>29.0</td>\n",
       "      <td>F2</td>\n",
       "      <td>2147</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>844</td>\n",
       "      <td>SPP1</td>\n",
       "      <td>6696</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>protein_protein</td>\n",
       "      <td>16.0</td>\n",
       "      <td>SNRPD2</td>\n",
       "      <td>6633</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>216</td>\n",
       "      <td>NCSTN</td>\n",
       "      <td>23385</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>protein_protein</td>\n",
       "      <td>33.0</td>\n",
       "      <td>GSK3B</td>\n",
       "      <td>2932</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>4817</td>\n",
       "      <td>NOTCH2</td>\n",
       "      <td>4853</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>protein_protein</td>\n",
       "      <td>33.0</td>\n",
       "      <td>GSK3B</td>\n",
       "      <td>2932</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>8918</td>\n",
       "      <td>ADAP1</td>\n",
       "      <td>11033</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>protein_protein</td>\n",
       "      <td>61.0</td>\n",
       "      <td>CALM1</td>\n",
       "      <td>801</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>2621</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>3845</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249262</th>\n",
       "      <td>249305</td>\n",
       "      <td>subordinate</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>Poverty</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>-1</td>\n",
       "      <td>economic stability</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249263</th>\n",
       "      <td>249306</td>\n",
       "      <td>subordinate</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>Access to Foods That Support Healthy Dietary P...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>-1</td>\n",
       "      <td>neighborhood and built environment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249264</th>\n",
       "      <td>249307</td>\n",
       "      <td>subordinate</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>Crime and Violence</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>-1</td>\n",
       "      <td>neighborhood and built environment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249265</th>\n",
       "      <td>249308</td>\n",
       "      <td>subordinate</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>Environmental Conditions</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>-1</td>\n",
       "      <td>neighborhood and built environment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249266</th>\n",
       "      <td>249309</td>\n",
       "      <td>subordinate</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>Quality of Housing</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH subtype</td>\n",
       "      <td>healthypeople</td>\n",
       "      <td>-1</td>\n",
       "      <td>neighborhood and built environment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>249267 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         index         relation  x_index  \\\n",
       "0            0  protein_protein     29.0   \n",
       "1            1  protein_protein     16.0   \n",
       "2            2  protein_protein     33.0   \n",
       "3            3  protein_protein     33.0   \n",
       "4            4  protein_protein     61.0   \n",
       "...        ...              ...      ...   \n",
       "249262  249305      subordinate     -1.0   \n",
       "249263  249306      subordinate     -1.0   \n",
       "249264  249307      subordinate     -1.0   \n",
       "249265  249308      subordinate     -1.0   \n",
       "249266  249309      subordinate     -1.0   \n",
       "\n",
       "                                                   x_name  x_id        x_type  \\\n",
       "0                                                      F2  2147  gene/protein   \n",
       "1                                                  SNRPD2  6633  gene/protein   \n",
       "2                                                   GSK3B  2932  gene/protein   \n",
       "3                                                   GSK3B  2932  gene/protein   \n",
       "4                                                   CALM1   801  gene/protein   \n",
       "...                                                   ...   ...           ...   \n",
       "249262                                            Poverty   NaN  SDoH subtype   \n",
       "249263  Access to Foods That Support Healthy Dietary P...   NaN  SDoH subtype   \n",
       "249264                                 Crime and Violence   NaN  SDoH subtype   \n",
       "249265                           Environmental Conditions   NaN  SDoH subtype   \n",
       "249266                                 Quality of Housing   NaN  SDoH subtype   \n",
       "\n",
       "             x_source  y_index                              y_name   y_id  \\\n",
       "0                NCBI      844                                SPP1   6696   \n",
       "1                NCBI      216                               NCSTN  23385   \n",
       "2                NCBI     4817                              NOTCH2   4853   \n",
       "3                NCBI     8918                               ADAP1  11033   \n",
       "4                NCBI     2621                                KRAS   3845   \n",
       "...               ...      ...                                 ...    ...   \n",
       "249262  healthypeople       -1                  economic stability    NaN   \n",
       "249263  healthypeople       -1  neighborhood and built environment    NaN   \n",
       "249264  healthypeople       -1  neighborhood and built environment    NaN   \n",
       "249265  healthypeople       -1  neighborhood and built environment    NaN   \n",
       "249266  healthypeople       -1  neighborhood and built environment    NaN   \n",
       "\n",
       "              y_type       y_source  \n",
       "0       gene/protein           NCBI  \n",
       "1       gene/protein           NCBI  \n",
       "2       gene/protein           NCBI  \n",
       "3       gene/protein           NCBI  \n",
       "4       gene/protein           NCBI  \n",
       "...              ...            ...  \n",
       "249262     SDoH type  healthypeople  \n",
       "249263     SDoH type  healthypeople  \n",
       "249264     SDoH type  healthypeople  \n",
       "249265     SDoH type  healthypeople  \n",
       "249266     SDoH type  healthypeople  \n",
       "\n",
       "[249267 rows x 12 columns]"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "testdf1 = alzheimer_sub[['relation', 'x_index', 'x_name', 'x_id', 'x_type', 'x_source', 'y_index', 'y_name', 'y_id', 'y_type', 'y_source']]\n",
    "testdf2 = SDoHKG[['relation', 'x_index', 'x_name', 'x_id', 'x_type', 'x_source', 'y_index', 'y_name', 'y_id', 'y_type', 'y_source']]\n",
    "testdf = pd.concat([testdf1, testdf2], axis=0).reset_index(drop=True)\n",
    "testdf.drop_duplicates(inplace=True)\n",
    "testdf.reset_index(inplace=True)\n",
    "testdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "testdf.to_csv('data/addsdoh.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#testdf = pd.read_csv('data/addsdoh.csv', low_memory=False)\n",
    "testdf = pd.read_csv('data/alzheimer_sub_graph.csv', low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>relation</th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_source</th>\n",
       "      <th>y_index</th>\n",
       "      <th>y_name</th>\n",
       "      <th>y_id</th>\n",
       "      <th>y_type</th>\n",
       "      <th>y_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8544</th>\n",
       "      <td>8547</td>\n",
       "      <td>drug_protein</td>\n",
       "      <td>15900.0</td>\n",
       "      <td>Manganese</td>\n",
       "      <td>DB06757</td>\n",
       "      <td>drug</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>4003</td>\n",
       "      <td>TF</td>\n",
       "      <td>7018</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186743</th>\n",
       "      <td>186764</td>\n",
       "      <td>exposure_protein</td>\n",
       "      <td>61759.0</td>\n",
       "      <td>Manganese</td>\n",
       "      <td>D008345</td>\n",
       "      <td>exposure</td>\n",
       "      <td>CTD</td>\n",
       "      <td>4003</td>\n",
       "      <td>TF</td>\n",
       "      <td>7018</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245265</th>\n",
       "      <td>245307</td>\n",
       "      <td>associates</td>\n",
       "      <td>15900.0</td>\n",
       "      <td>Manganese</td>\n",
       "      <td>DB06757</td>\n",
       "      <td>drug</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>-1</td>\n",
       "      <td>occupational</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247283</th>\n",
       "      <td>247325</td>\n",
       "      <td>is naturally found in</td>\n",
       "      <td>15900.0</td>\n",
       "      <td>Manganese</td>\n",
       "      <td>DB06757</td>\n",
       "      <td>drug</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>-1</td>\n",
       "      <td>environment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>healthypeople</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         index               relation  x_index     x_name     x_id    x_type  \\\n",
       "8544      8547           drug_protein  15900.0  Manganese  DB06757      drug   \n",
       "186743  186764       exposure_protein  61759.0  Manganese  D008345  exposure   \n",
       "245265  245307             associates  15900.0  Manganese  DB06757      drug   \n",
       "247283  247325  is naturally found in  15900.0  Manganese  DB06757      drug   \n",
       "\n",
       "        x_source  y_index        y_name  y_id        y_type       y_source  \n",
       "8544    DrugBank     4003            TF  7018  gene/protein           NCBI  \n",
       "186743       CTD     4003            TF  7018  gene/protein           NCBI  \n",
       "245265  DrugBank       -1  occupational   NaN   SDoH factor  healthypeople  \n",
       "247283  DrugBank       -1   environment   NaN   SDoH factor  healthypeople  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "testdf[testdf['x_name']=='Manganese']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "t1 = testdf[['x_index', 'x_name', 'x_id', 'x_type', 'x_source']]\n",
    "t2 = testdf[['y_index', 'y_name', 'y_id', 'y_type', 'y_source']]\n",
    "t2.columns = ['x_index', 'x_name', 'x_id', 'x_type', 'x_source']\n",
    "# concat pkg1 and pkg2\n",
    "testid = pd.concat([t1, t2], axis=0)\n",
    "testid = testid.drop_duplicates(subset=('x_name', 'x_type'))\n",
    "#testid = testid.drop_duplicates(subset='x_name')\n",
    "testid.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>F2</td>\n",
       "      <td>2147</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>SNRPD2</td>\n",
       "      <td>6633</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>GSK3B</td>\n",
       "      <td>2932</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>CALM1</td>\n",
       "      <td>801</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>INS</td>\n",
       "      <td>3630</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19185</th>\n",
       "      <td>0</td>\n",
       "      <td>healthcare access and quality</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19186</th>\n",
       "      <td>1</td>\n",
       "      <td>education access and quality</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19187</th>\n",
       "      <td>2</td>\n",
       "      <td>social and community context</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19188</th>\n",
       "      <td>3</td>\n",
       "      <td>economic stability</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19189</th>\n",
       "      <td>4</td>\n",
       "      <td>neighborhood and built environment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SDoH type</td>\n",
       "      <td>healthypeople</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19190 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       x_index                              x_name  x_id        x_type  \\\n",
       "0            0                                  F2  2147  gene/protein   \n",
       "1            1                              SNRPD2  6633  gene/protein   \n",
       "2            2                               GSK3B  2932  gene/protein   \n",
       "3            3                               CALM1   801  gene/protein   \n",
       "4            4                                 INS  3630  gene/protein   \n",
       "...        ...                                 ...   ...           ...   \n",
       "19185        0       healthcare access and quality   NaN     SDoH type   \n",
       "19186        1        education access and quality   NaN     SDoH type   \n",
       "19187        2        social and community context   NaN     SDoH type   \n",
       "19188        3                  economic stability   NaN     SDoH type   \n",
       "19189        4  neighborhood and built environment   NaN     SDoH type   \n",
       "\n",
       "            x_source  \n",
       "0               NCBI  \n",
       "1               NCBI  \n",
       "2               NCBI  \n",
       "3               NCBI  \n",
       "4               NCBI  \n",
       "...              ...  \n",
       "19185  healthypeople  \n",
       "19186  healthypeople  \n",
       "19187  healthypeople  \n",
       "19188  healthypeople  \n",
       "19189  healthypeople  \n",
       "\n",
       "[19190 rows x 5 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "testid['x_index'] = testid.groupby('x_type').cumcount().astype(int)  \n",
    "testid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2179</th>\n",
       "      <td>972</td>\n",
       "      <td>Manganese</td>\n",
       "      <td>DB06757</td>\n",
       "      <td>drug</td>\n",
       "      <td>DrugBank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9388</th>\n",
       "      <td>29</td>\n",
       "      <td>Manganese</td>\n",
       "      <td>D008345</td>\n",
       "      <td>exposure</td>\n",
       "      <td>CTD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      x_index     x_name     x_id    x_type  x_source\n",
       "2179      972  Manganese  DB06757      drug  DrugBank\n",
       "9388       29  Manganese  D008345  exposure       CTD"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find 'x_name' = manganese in testid\n",
    "testid[testid['x_name']=='Manganese']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "testid.to_csv('data/pkgid.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['APOE', 'TREM2', 'CD33', 'CLU', 'ABCA7', 'BIN1', 'CR1', 'SORL1', 'MS4A gene cluster', 'CD2AP', 'EPHA1', 'INPP5D', 'SHARPIN', 'TREML2', 'HLA-DR', 'ABI3', 'PLCG2', 'MEF2C', 'PICALM', 'C7', 'CASS4', 'SPI1', 'PTK2B', 'ADAMTS1', 'IGHG3', 'CLNK']\n",
    "for i in genelist:\n",
    "    print(testid[testid['x_name'] == i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "id_type = testid['x_type'].unique() \n",
    "dfs_by_type = {x_type: df_sub for x_type, df_sub in testid.groupby('x_type')}\n",
    "\n",
    "for i in range(len(id_type)):\n",
    "    subdf = dfs_by_type.get(id_type[i])\n",
    "    subdf = subdf[['x_index']]\n",
    "    subdf.columns = ['node_id']\n",
    "    #path = 'data/rgcn/merge/nodes_'+str(i)+ '.csv'\n",
    "    path = 'data/rgcn/pkg/nodes_'+str(i)+ '.csv'\n",
    "    subdf.to_csv(path, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['gene/protein', 'drug', 'disease', 'effect/phenotype', 'exposure',\n",
       "       'molecular_function', 'cellular_component', 'biological_process',\n",
       "       'pathway', 'anatomy', 'Species', 'Variant', 'CellLine',\n",
       "       'SDoH factor', 'SDoH subtype', 'SDoH type'], dtype=object)"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relation</th>\n",
       "      <th>display_relation</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_source</th>\n",
       "      <th>y_id</th>\n",
       "      <th>y_type</th>\n",
       "      <th>y_name</th>\n",
       "      <th>y_source</th>\n",
       "      <th>x_index</th>\n",
       "      <th>y_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>2147</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>F2</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>6696</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>SPP1</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>0</td>\n",
       "      <td>1119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>6633</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>SNRPD2</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>23385</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NCSTN</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>2932</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>GSK3B</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>4853</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>NOTCH2</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>2</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>2932</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>GSK3B</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>11033</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>ADAP1</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>2</td>\n",
       "      <td>1436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>protein_protein</td>\n",
       "      <td>ppi</td>\n",
       "      <td>801</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>CALM1</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>3845</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>3</td>\n",
       "      <td>1437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245247</th>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>expression absent</td>\n",
       "      <td>81</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>metanephros</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>7124</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>TNF</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>108</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245248</th>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>expression absent</td>\n",
       "      <td>1377</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>quadriceps femoris</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>7124</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>TNF</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>110</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245249</th>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>expression absent</td>\n",
       "      <td>1379</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>vastus lateralis</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>3630</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>INS</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>111</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245250</th>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>expression absent</td>\n",
       "      <td>1507</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>biceps brachii</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>3630</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>INS</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>113</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245251</th>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>expression absent</td>\n",
       "      <td>2349</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>myocardium</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>3630</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>INS</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>116</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>245252 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      relation   display_relation  x_id        x_type  \\\n",
       "0              protein_protein                ppi  2147  gene/protein   \n",
       "1              protein_protein                ppi  6633  gene/protein   \n",
       "2              protein_protein                ppi  2932  gene/protein   \n",
       "3              protein_protein                ppi  2932  gene/protein   \n",
       "4              protein_protein                ppi   801  gene/protein   \n",
       "...                        ...                ...   ...           ...   \n",
       "245247  anatomy_protein_absent  expression absent    81       anatomy   \n",
       "245248  anatomy_protein_absent  expression absent  1377       anatomy   \n",
       "245249  anatomy_protein_absent  expression absent  1379       anatomy   \n",
       "245250  anatomy_protein_absent  expression absent  1507       anatomy   \n",
       "245251  anatomy_protein_absent  expression absent  2349       anatomy   \n",
       "\n",
       "                    x_name x_source   y_id        y_type  y_name y_source  \\\n",
       "0                       F2     NCBI   6696  gene/protein    SPP1     NCBI   \n",
       "1                   SNRPD2     NCBI  23385  gene/protein   NCSTN     NCBI   \n",
       "2                    GSK3B     NCBI   4853  gene/protein  NOTCH2     NCBI   \n",
       "3                    GSK3B     NCBI  11033  gene/protein   ADAP1     NCBI   \n",
       "4                    CALM1     NCBI   3845  gene/protein    KRAS     NCBI   \n",
       "...                    ...      ...    ...           ...     ...      ...   \n",
       "245247         metanephros   UBERON   7124  gene/protein     TNF     NCBI   \n",
       "245248  quadriceps femoris   UBERON   7124  gene/protein     TNF     NCBI   \n",
       "245249    vastus lateralis   UBERON   3630  gene/protein     INS     NCBI   \n",
       "245250      biceps brachii   UBERON   3630  gene/protein     INS     NCBI   \n",
       "245251          myocardium   UBERON   3630  gene/protein     INS     NCBI   \n",
       "\n",
       "        x_index  y_index  \n",
       "0             0     1119  \n",
       "1             1        5  \n",
       "2             2      120  \n",
       "3             2     1436  \n",
       "4             3     1437  \n",
       "...         ...      ...  \n",
       "245247      108       40  \n",
       "245248      110       40  \n",
       "245249      111        4  \n",
       "245250      113        4  \n",
       "245251      116        4  \n",
       "\n",
       "[245252 rows x 12 columns]"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "testdf = testdf.merge(testid[['x_name', 'x_type', 'x_index']], how='left', left_on=['x_name', 'x_type'], right_on=['x_name', 'x_type'])\n",
    "testdf = testdf.drop(columns=['x_index_x'])\n",
    "testdf.rename(columns={'x_index_y': 'x_index'}, inplace=True)\n",
    "\n",
    "# Merge to replace y_index\n",
    "testdf = testdf.merge(testid[['x_name', 'x_type', 'x_index']], how='left', left_on=['y_name', 'y_type'], right_on=['x_name', 'x_type'])\n",
    "testdf = testdf.drop(columns=['x_name_y', 'x_type_y', 'y_index'])\n",
    "testdf.rename(columns={'x_index_y': 'y_index'}, inplace=True)\n",
    "testdf.rename(columns={'x_index_x': 'x_index'}, inplace=True)\n",
    "testdf.rename(columns={'x_name_x': 'x_name'}, inplace=True)\n",
    "testdf.rename(columns={'x_type_x': 'x_type'}, inplace=True)\n",
    "\n",
    "testdf "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relation</th>\n",
       "      <th>display_relation</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_source</th>\n",
       "      <th>y_id</th>\n",
       "      <th>y_type</th>\n",
       "      <th>y_name</th>\n",
       "      <th>y_source</th>\n",
       "      <th>x_index</th>\n",
       "      <th>y_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8547</th>\n",
       "      <td>drug_protein</td>\n",
       "      <td>target</td>\n",
       "      <td>DB06757</td>\n",
       "      <td>drug</td>\n",
       "      <td>Manganese</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>7018</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>TF</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>972</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186764</th>\n",
       "      <td>exposure_protein</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>D008345</td>\n",
       "      <td>exposure</td>\n",
       "      <td>Manganese</td>\n",
       "      <td>CTD</td>\n",
       "      <td>7018</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>TF</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>29</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204981</th>\n",
       "      <td>drug_protein</td>\n",
       "      <td>target</td>\n",
       "      <td>7018</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>TF</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>DB06757</td>\n",
       "      <td>drug</td>\n",
       "      <td>Manganese</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>79</td>\n",
       "      <td>972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235582</th>\n",
       "      <td>exposure_protein</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>7018</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>TF</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>D008345</td>\n",
       "      <td>exposure</td>\n",
       "      <td>Manganese</td>\n",
       "      <td>CTD</td>\n",
       "      <td>79</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                relation display_relation     x_id        x_type     x_name  \\\n",
       "8547        drug_protein           target  DB06757          drug  Manganese   \n",
       "186764  exposure_protein   interacts with  D008345      exposure  Manganese   \n",
       "204981      drug_protein           target     7018  gene/protein         TF   \n",
       "235582  exposure_protein   interacts with     7018  gene/protein         TF   \n",
       "\n",
       "        x_source     y_id        y_type     y_name  y_source  x_index  y_index  \n",
       "8547    DrugBank     7018  gene/protein         TF      NCBI      972       79  \n",
       "186764       CTD     7018  gene/protein         TF      NCBI       29       79  \n",
       "204981      NCBI  DB06757          drug  Manganese  DrugBank       79      972  \n",
       "235582      NCBI  D008345      exposure  Manganese       CTD       79       29  "
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "testdf [(testdf['x_name']=='Manganese')|(testdf['y_name']=='Manganese')] \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_type</th>\n",
       "      <th>relation</th>\n",
       "      <th>y_index</th>\n",
       "      <th>y_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>protein_protein</td>\n",
       "      <td>1119</td>\n",
       "      <td>gene/protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>protein_protein</td>\n",
       "      <td>5</td>\n",
       "      <td>gene/protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>protein_protein</td>\n",
       "      <td>120</td>\n",
       "      <td>gene/protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>protein_protein</td>\n",
       "      <td>1436</td>\n",
       "      <td>gene/protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>protein_protein</td>\n",
       "      <td>1437</td>\n",
       "      <td>gene/protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245247</th>\n",
       "      <td>108</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>40</td>\n",
       "      <td>gene/protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245248</th>\n",
       "      <td>110</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>40</td>\n",
       "      <td>gene/protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245249</th>\n",
       "      <td>111</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>4</td>\n",
       "      <td>gene/protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245250</th>\n",
       "      <td>113</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>4</td>\n",
       "      <td>gene/protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245251</th>\n",
       "      <td>116</td>\n",
       "      <td>anatomy</td>\n",
       "      <td>anatomy_protein_absent</td>\n",
       "      <td>4</td>\n",
       "      <td>gene/protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>245252 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        x_index        x_type                relation  y_index        y_type\n",
       "0             0  gene/protein         protein_protein     1119  gene/protein\n",
       "1             1  gene/protein         protein_protein        5  gene/protein\n",
       "2             2  gene/protein         protein_protein      120  gene/protein\n",
       "3             2  gene/protein         protein_protein     1436  gene/protein\n",
       "4             3  gene/protein         protein_protein     1437  gene/protein\n",
       "...         ...           ...                     ...      ...           ...\n",
       "245247      108       anatomy  anatomy_protein_absent       40  gene/protein\n",
       "245248      110       anatomy  anatomy_protein_absent       40  gene/protein\n",
       "245249      111       anatomy  anatomy_protein_absent        4  gene/protein\n",
       "245250      113       anatomy  anatomy_protein_absent        4  gene/protein\n",
       "245251      116       anatomy  anatomy_protein_absent        4  gene/protein\n",
       "\n",
       "[245252 rows x 5 columns]"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t3 = testdf[['x_index','x_type','relation', 'y_index','y_type']] \n",
    "t3 = t3.dropna(subset=['x_index', 'y_index'])\n",
    "t3['x_index'] = t3['x_index'].astype(int)\n",
    "t3['y_index'] = t3['y_index'].astype(int)\n",
    "t3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_type</th>\n",
       "      <th>relation</th>\n",
       "      <th>y_index</th>\n",
       "      <th>y_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3207</th>\n",
       "      <td>9</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>protein_protein</td>\n",
       "      <td>972</td>\n",
       "      <td>gene/protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5140</th>\n",
       "      <td>972</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>protein_protein</td>\n",
       "      <td>79</td>\n",
       "      <td>gene/protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8544</th>\n",
       "      <td>972</td>\n",
       "      <td>drug</td>\n",
       "      <td>drug_protein</td>\n",
       "      <td>79</td>\n",
       "      <td>gene/protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163623</th>\n",
       "      <td>972</td>\n",
       "      <td>disease</td>\n",
       "      <td>disease_phenotype_positive</td>\n",
       "      <td>129</td>\n",
       "      <td>effect/phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163624</th>\n",
       "      <td>972</td>\n",
       "      <td>disease</td>\n",
       "      <td>disease_phenotype_positive</td>\n",
       "      <td>131</td>\n",
       "      <td>effect/phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174574</th>\n",
       "      <td>96</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>disease_protein</td>\n",
       "      <td>972</td>\n",
       "      <td>disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185697</th>\n",
       "      <td>860</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>bioprocess_protein</td>\n",
       "      <td>972</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199620</th>\n",
       "      <td>972</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>protein_protein</td>\n",
       "      <td>9</td>\n",
       "      <td>gene/protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201553</th>\n",
       "      <td>79</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>protein_protein</td>\n",
       "      <td>972</td>\n",
       "      <td>gene/protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204957</th>\n",
       "      <td>79</td>\n",
       "      <td>gene/protein</td>\n",
       "      <td>drug_protein</td>\n",
       "      <td>972</td>\n",
       "      <td>drug</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212422</th>\n",
       "      <td>129</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>disease_phenotype_positive</td>\n",
       "      <td>972</td>\n",
       "      <td>disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212423</th>\n",
       "      <td>131</td>\n",
       "      <td>effect/phenotype</td>\n",
       "      <td>disease_phenotype_positive</td>\n",
       "      <td>972</td>\n",
       "      <td>disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223373</th>\n",
       "      <td>972</td>\n",
       "      <td>disease</td>\n",
       "      <td>disease_protein</td>\n",
       "      <td>96</td>\n",
       "      <td>gene/protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234494</th>\n",
       "      <td>972</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>bioprocess_protein</td>\n",
       "      <td>860</td>\n",
       "      <td>gene/protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245265</th>\n",
       "      <td>972</td>\n",
       "      <td>drug</td>\n",
       "      <td>associates</td>\n",
       "      <td>38</td>\n",
       "      <td>SDoH factor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247283</th>\n",
       "      <td>972</td>\n",
       "      <td>drug</td>\n",
       "      <td>is naturally found in</td>\n",
       "      <td>81</td>\n",
       "      <td>SDoH factor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248916</th>\n",
       "      <td>972</td>\n",
       "      <td>SDoH factor</td>\n",
       "      <td>subordinate</td>\n",
       "      <td>0</td>\n",
       "      <td>SDoH subtype</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        x_index              x_type                    relation  y_index  \\\n",
       "3207          9        gene/protein             protein_protein      972   \n",
       "5140        972        gene/protein             protein_protein       79   \n",
       "8544        972                drug                drug_protein       79   \n",
       "163623      972             disease  disease_phenotype_positive      129   \n",
       "163624      972             disease  disease_phenotype_positive      131   \n",
       "174574       96        gene/protein             disease_protein      972   \n",
       "185697      860        gene/protein          bioprocess_protein      972   \n",
       "199620      972        gene/protein             protein_protein        9   \n",
       "201553       79        gene/protein             protein_protein      972   \n",
       "204957       79        gene/protein                drug_protein      972   \n",
       "212422      129    effect/phenotype  disease_phenotype_positive      972   \n",
       "212423      131    effect/phenotype  disease_phenotype_positive      972   \n",
       "223373      972             disease             disease_protein       96   \n",
       "234494      972  biological_process          bioprocess_protein      860   \n",
       "245265      972                drug                  associates       38   \n",
       "247283      972                drug       is naturally found in       81   \n",
       "248916      972         SDoH factor                 subordinate        0   \n",
       "\n",
       "                    y_type  \n",
       "3207          gene/protein  \n",
       "5140          gene/protein  \n",
       "8544          gene/protein  \n",
       "163623    effect/phenotype  \n",
       "163624    effect/phenotype  \n",
       "174574             disease  \n",
       "185697  biological_process  \n",
       "199620        gene/protein  \n",
       "201553        gene/protein  \n",
       "204957                drug  \n",
       "212422             disease  \n",
       "212423             disease  \n",
       "223373        gene/protein  \n",
       "234494        gene/protein  \n",
       "245265         SDoH factor  \n",
       "247283         SDoH factor  \n",
       "248916        SDoH subtype  "
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t3 [(t3 ['y_index'] == 972)|(t3 ['x_index'] == 972)] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gene/protein': 0,\n",
       " 'drug': 1,\n",
       " 'disease': 2,\n",
       " 'effect/phenotype': 3,\n",
       " 'exposure': 4,\n",
       " 'molecular_function': 5,\n",
       " 'cellular_component': 6,\n",
       " 'biological_process': 7,\n",
       " 'pathway': 8,\n",
       " 'anatomy': 9,\n",
       " 'Species': 10,\n",
       " 'Variant': 11,\n",
       " 'CellLine': 12,\n",
       " 'SDoH factor': 13,\n",
       " 'SDoH subtype': 14,\n",
       " 'SDoH type': 15}"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_dict = {value: index for index, value in enumerate(id_type)}\n",
    "mapping_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sub-DataFrame for x_type = anatomy and y_type = gene/protein:\n",
      "        x_index   x_type                 relation  y_index        y_type\n",
      "236515        0  anatomy  anatomy_protein_present      104  gene/protein\n",
      "236516        1  anatomy  anatomy_protein_present      104  gene/protein\n",
      "236517        2  anatomy  anatomy_protein_present      104  gene/protein\n",
      "236518        3  anatomy  anatomy_protein_present      104  gene/protein\n",
      "236519        4  anatomy  anatomy_protein_present      104  gene/protein\n",
      "...         ...      ...                      ...      ...           ...\n",
      "245247      108  anatomy   anatomy_protein_absent       40  gene/protein\n",
      "245248      110  anatomy   anatomy_protein_absent       40  gene/protein\n",
      "245249      111  anatomy   anatomy_protein_absent        4  gene/protein\n",
      "245250      113  anatomy   anatomy_protein_absent        4  gene/protein\n",
      "245251      116  anatomy   anatomy_protein_absent        4  gene/protein\n",
      "\n",
      "[8737 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = biological_process and y_type = gene/protein:\n",
      "        x_index              x_type            relation  y_index        y_type\n",
      "232452        0  biological_process  bioprocess_protein      205  gene/protein\n",
      "232453        0  biological_process  bioprocess_protein       21  gene/protein\n",
      "232454        0  biological_process  bioprocess_protein      640  gene/protein\n",
      "232455        0  biological_process  bioprocess_protein      112  gene/protein\n",
      "232456        0  biological_process  bioprocess_protein       36  gene/protein\n",
      "...         ...                 ...                 ...      ...           ...\n",
      "235489     1783  biological_process  bioprocess_protein     1421  gene/protein\n",
      "235490     1784  biological_process  bioprocess_protein     1421  gene/protein\n",
      "235491     1785  biological_process  bioprocess_protein     1421  gene/protein\n",
      "235492     1786  biological_process  bioprocess_protein     1421  gene/protein\n",
      "235493     1787  biological_process  bioprocess_protein     1426  gene/protein\n",
      "\n",
      "[3042 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = cellular_component and y_type = gene/protein:\n",
      "        x_index              x_type          relation  y_index        y_type\n",
      "231437        0  cellular_component  cellcomp_protein      112  gene/protein\n",
      "231438        1  cellular_component  cellcomp_protein       22  gene/protein\n",
      "231439        1  cellular_component  cellcomp_protein      223  gene/protein\n",
      "231440        1  cellular_component  cellcomp_protein       75  gene/protein\n",
      "231441        1  cellular_component  cellcomp_protein       11  gene/protein\n",
      "...         ...                 ...               ...      ...           ...\n",
      "232447      266  cellular_component  cellcomp_protein       17  gene/protein\n",
      "232448      267  cellular_component  cellcomp_protein      107  gene/protein\n",
      "232449      268  cellular_component  cellcomp_protein       91  gene/protein\n",
      "232450      269  cellular_component  cellcomp_protein       48  gene/protein\n",
      "232451      270  cellular_component  cellcomp_protein      139  gene/protein\n",
      "\n",
      "[1015 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = disease and y_type = disease:\n",
      "        x_index   x_type         relation  y_index   y_type\n",
      "175128     4242  disease  disease_disease     5138  disease\n",
      "175129     4242  disease  disease_disease      739  disease\n",
      "175130     4242  disease  disease_disease     5228  disease\n",
      "175131     4243  disease  disease_disease     5229  disease\n",
      "175132     4244  disease  disease_disease     4245  disease\n",
      "...         ...      ...              ...      ...      ...\n",
      "224238      377  disease  disease_disease     4245  disease\n",
      "224239     5298  disease  disease_disease     4245  disease\n",
      "224240     5299  disease  disease_disease     4242  disease\n",
      "224241      335  disease  disease_disease     4242  disease\n",
      "224242      293  disease  disease_disease     4242  disease\n",
      "\n",
      "[592 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = disease and y_type = drug:\n",
      "        x_index   x_type          relation  y_index y_type\n",
      "205857     4254  disease  contraindication     1253   drug\n",
      "205858     4255  disease  contraindication     1253   drug\n",
      "205859     4254  disease  contraindication      756   drug\n",
      "205860     4255  disease  contraindication      756   drug\n",
      "205861     4254  disease  contraindication     1254   drug\n",
      "...         ...      ...               ...      ...    ...\n",
      "210207     3191  disease        indication       19   drug\n",
      "210208     4620  disease        indication       19   drug\n",
      "210209     4619  disease        indication       18   drug\n",
      "210210     3191  disease        indication       18   drug\n",
      "210211     4620  disease        indication       18   drug\n",
      "\n",
      "[4355 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = disease and y_type = effect/phenotype:\n",
      "        x_index   x_type                    relation  y_index  \\\n",
      "161933        0  disease  disease_phenotype_negative      117   \n",
      "161934        1  disease  disease_phenotype_negative      117   \n",
      "161935        2  disease  disease_phenotype_negative      117   \n",
      "161936        3  disease  disease_phenotype_negative      117   \n",
      "161937        4  disease  disease_phenotype_negative      117   \n",
      "...         ...      ...                         ...      ...   \n",
      "220027     4626  disease  disease_phenotype_positive        3   \n",
      "220028     4627  disease  disease_phenotype_positive        0   \n",
      "220029     4628  disease  disease_phenotype_positive        4   \n",
      "220030     4629  disease  disease_phenotype_positive        3   \n",
      "220031     4630  disease  disease_phenotype_positive        2   \n",
      "\n",
      "                  y_type  \n",
      "161933  effect/phenotype  \n",
      "161934  effect/phenotype  \n",
      "161935  effect/phenotype  \n",
      "161936  effect/phenotype  \n",
      "161937  effect/phenotype  \n",
      "...                  ...  \n",
      "220027  effect/phenotype  \n",
      "220028  effect/phenotype  \n",
      "220029  effect/phenotype  \n",
      "220030  effect/phenotype  \n",
      "220031  effect/phenotype  \n",
      "\n",
      "[9279 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = disease and y_type = exposure:\n",
      "        x_index   x_type          relation  y_index    y_type\n",
      "235810     4247  disease  exposure_disease        5  exposure\n",
      "235811      896  disease  exposure_disease       60  exposure\n",
      "235812     4310  disease  exposure_disease       60  exposure\n",
      "235813     5300  disease  exposure_disease       60  exposure\n",
      "235814       30  disease  exposure_disease       60  exposure\n",
      "...         ...      ...               ...      ...       ...\n",
      "236030     4834  disease  exposure_disease       51  exposure\n",
      "236031     4843  disease  exposure_disease       51  exposure\n",
      "236032     1186  disease  exposure_disease       51  exposure\n",
      "236033     4247  disease  exposure_disease       53  exposure\n",
      "236034     4247  disease  exposure_disease       56  exposure\n",
      "\n",
      "[225 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = disease and y_type = gene/protein:\n",
      "        x_index   x_type         relation  y_index        y_type\n",
      "220032      543  disease  disease_protein      223  gene/protein\n",
      "220033      543  disease  disease_protein       75  gene/protein\n",
      "220034      543  disease  disease_protein       43  gene/protein\n",
      "220035      543  disease  disease_protein      152  gene/protein\n",
      "220036      543  disease  disease_protein       51  gene/protein\n",
      "...         ...      ...              ...      ...           ...\n",
      "223943     4315  disease  disease_protein       40  gene/protein\n",
      "223944     4304  disease  disease_protein        4  gene/protein\n",
      "223945     5226  disease  disease_protein       94  gene/protein\n",
      "223946     4393  disease  disease_protein      532  gene/protein\n",
      "223947     5227  disease  disease_protein       96  gene/protein\n",
      "\n",
      "[3916 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = drug and y_type = disease:\n",
      "       x_index x_type          relation  y_index   y_type\n",
      "9423      1253   drug  contraindication     4254  disease\n",
      "9424      1253   drug  contraindication     4255  disease\n",
      "9425       756   drug  contraindication     4254  disease\n",
      "9426       756   drug  contraindication     4255  disease\n",
      "9427      1254   drug  contraindication     4254  disease\n",
      "...        ...    ...               ...      ...      ...\n",
      "13773       19   drug        indication     3191  disease\n",
      "13774       19   drug        indication     4620  disease\n",
      "13775       18   drug        indication     4619  disease\n",
      "13776       18   drug        indication     3191  disease\n",
      "13777       18   drug        indication     4620  disease\n",
      "\n",
      "[4355 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = drug and y_type = drug:\n",
      "        x_index x_type   relation  y_index y_type\n",
      "13778       981   drug  drug_drug     1785   drug\n",
      "13779        63   drug  drug_drug     1785   drug\n",
      "13780        12   drug  drug_drug     1785   drug\n",
      "13781        13   drug  drug_drug     1785   drug\n",
      "13782        14   drug  drug_drug     1785   drug\n",
      "...         ...    ...        ...      ...    ...\n",
      "161387      149   drug  drug_drug     3536   drug\n",
      "161388      149   drug  drug_drug     3537   drug\n",
      "161389      149   drug  drug_drug     3541   drug\n",
      "161390      149   drug  drug_drug     3542   drug\n",
      "161391      149   drug  drug_drug      423   drug\n",
      "\n",
      "[147614 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = drug and y_type = effect/phenotype:\n",
      "        x_index x_type     relation  y_index            y_type\n",
      "175425     1142   drug  drug_effect      111  effect/phenotype\n",
      "175426     1142   drug  drug_effect        7  effect/phenotype\n",
      "175427     3624   drug  drug_effect      142  effect/phenotype\n",
      "175428     3624   drug  drug_effect        7  effect/phenotype\n",
      "175429     3447   drug  drug_effect        7  effect/phenotype\n",
      "...         ...    ...          ...      ...               ...\n",
      "181854     2306   drug  drug_effect        7  effect/phenotype\n",
      "181855     2306   drug  drug_effect        0  effect/phenotype\n",
      "181856     2306   drug  drug_effect      132  effect/phenotype\n",
      "181857     2306   drug  drug_effect      123  effect/phenotype\n",
      "181858     2306   drug  drug_effect      133  effect/phenotype\n",
      "\n",
      "[6434 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = drug and y_type = gene/protein:\n",
      "      x_index x_type      relation  y_index        y_type\n",
      "7106        0   drug  drug_protein       79  gene/protein\n",
      "7107        1   drug  drug_protein       79  gene/protein\n",
      "7108        2   drug  drug_protein       79  gene/protein\n",
      "7109        3   drug  drug_protein       79  gene/protein\n",
      "7110        4   drug  drug_protein       79  gene/protein\n",
      "...       ...    ...           ...      ...           ...\n",
      "9418      776   drug  drug_protein     4715  gene/protein\n",
      "9419       66   drug  drug_protein       28  gene/protein\n",
      "9420       67   drug  drug_protein       28  gene/protein\n",
      "9421       66   drug  drug_protein      653  gene/protein\n",
      "9422       67   drug  drug_protein      653  gene/protein\n",
      "\n",
      "[2317 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = effect/phenotype and y_type = disease:\n",
      "        x_index            x_type                    relation  y_index  \\\n",
      "171201        0  effect/phenotype  disease_phenotype_positive     4621   \n",
      "171202        1  effect/phenotype  disease_phenotype_positive     4622   \n",
      "171203        2  effect/phenotype  disease_phenotype_positive      896   \n",
      "171204        3  effect/phenotype  disease_phenotype_positive     4623   \n",
      "171205        0  effect/phenotype  disease_phenotype_positive     4624   \n",
      "...         ...               ...                         ...      ...   \n",
      "220016      116  effect/phenotype  disease_phenotype_positive     1822   \n",
      "220017        7  effect/phenotype  disease_phenotype_positive     4240   \n",
      "220018      117  effect/phenotype  disease_phenotype_positive     2245   \n",
      "220019      140  effect/phenotype  disease_phenotype_positive        8   \n",
      "220020      151  effect/phenotype  disease_phenotype_positive     4241   \n",
      "\n",
      "         y_type  \n",
      "171201  disease  \n",
      "171202  disease  \n",
      "171203  disease  \n",
      "171204  disease  \n",
      "171205  disease  \n",
      "...         ...  \n",
      "220016  disease  \n",
      "220017  disease  \n",
      "220018  disease  \n",
      "220019  disease  \n",
      "220020  disease  \n",
      "\n",
      "[9279 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = effect/phenotype and y_type = drug:\n",
      "        x_index            x_type     relation  y_index y_type\n",
      "224243      111  effect/phenotype  drug_effect     1142   drug\n",
      "224244        7  effect/phenotype  drug_effect     1142   drug\n",
      "224245      142  effect/phenotype  drug_effect     3624   drug\n",
      "224246        7  effect/phenotype  drug_effect     3624   drug\n",
      "224247        7  effect/phenotype  drug_effect     3447   drug\n",
      "...         ...               ...          ...      ...    ...\n",
      "230672        7  effect/phenotype  drug_effect     2306   drug\n",
      "230673        0  effect/phenotype  drug_effect     2306   drug\n",
      "230674      132  effect/phenotype  drug_effect     2306   drug\n",
      "230675      123  effect/phenotype  drug_effect     2306   drug\n",
      "230676      133  effect/phenotype  drug_effect     2306   drug\n",
      "\n",
      "[6434 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = effect/phenotype and y_type = gene/protein:\n",
      "        x_index            x_type           relation  y_index        y_type\n",
      "210212        5  effect/phenotype  phenotype_protein      540  gene/protein\n",
      "210213        5  effect/phenotype  phenotype_protein       94  gene/protein\n",
      "210214        6  effect/phenotype  phenotype_protein       19  gene/protein\n",
      "210215        7  effect/phenotype  phenotype_protein     1207  gene/protein\n",
      "210216        7  effect/phenotype  phenotype_protein     1208  gene/protein\n",
      "...         ...               ...                ...      ...           ...\n",
      "210748      112  effect/phenotype  phenotype_protein     1414  gene/protein\n",
      "210749      113  effect/phenotype  phenotype_protein      108  gene/protein\n",
      "210750      114  effect/phenotype  phenotype_protein      697  gene/protein\n",
      "210751      115  effect/phenotype  phenotype_protein     1409  gene/protein\n",
      "210752      116  effect/phenotype  phenotype_protein     1415  gene/protein\n",
      "\n",
      "[541 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = exposure and y_type = disease:\n",
      "        x_index    x_type          relation  y_index   y_type\n",
      "186992        5  exposure  exposure_disease     4247  disease\n",
      "186993       60  exposure  exposure_disease      896  disease\n",
      "186994       60  exposure  exposure_disease     4310  disease\n",
      "186995       60  exposure  exposure_disease     5300  disease\n",
      "186996       60  exposure  exposure_disease       30  disease\n",
      "...         ...       ...               ...      ...      ...\n",
      "187212       51  exposure  exposure_disease     4834  disease\n",
      "187213       51  exposure  exposure_disease     4843  disease\n",
      "187214       51  exposure  exposure_disease     1186  disease\n",
      "187215       53  exposure  exposure_disease     4247  disease\n",
      "187216       56  exposure  exposure_disease     4247  disease\n",
      "\n",
      "[225 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = exposure and y_type = gene/protein:\n",
      "        x_index    x_type          relation  y_index        y_type\n",
      "186676        0  exposure  exposure_protein      152  gene/protein\n",
      "186677        1  exposure  exposure_protein      291  gene/protein\n",
      "186678        2  exposure  exposure_protein      152  gene/protein\n",
      "186679        3  exposure  exposure_protein       40  gene/protein\n",
      "186680        4  exposure  exposure_protein       96  gene/protein\n",
      "...         ...       ...               ...      ...           ...\n",
      "186987       56  exposure  exposure_protein       40  gene/protein\n",
      "186988       57  exposure  exposure_protein       19  gene/protein\n",
      "186989       58  exposure  exposure_protein       19  gene/protein\n",
      "186990       58  exposure  exposure_protein       40  gene/protein\n",
      "186991       59  exposure  exposure_protein       40  gene/protein\n",
      "\n",
      "[316 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = gene/protein and y_type = anatomy:\n",
      "        x_index        x_type                 relation  y_index   y_type\n",
      "187697      104  gene/protein  anatomy_protein_present        0  anatomy\n",
      "187698      104  gene/protein  anatomy_protein_present        1  anatomy\n",
      "187699      104  gene/protein  anatomy_protein_present        2  anatomy\n",
      "187700      104  gene/protein  anatomy_protein_present        3  anatomy\n",
      "187701      104  gene/protein  anatomy_protein_present        4  anatomy\n",
      "...         ...           ...                      ...      ...      ...\n",
      "196429       40  gene/protein   anatomy_protein_absent      108  anatomy\n",
      "196430       40  gene/protein   anatomy_protein_absent      110  anatomy\n",
      "196431        4  gene/protein   anatomy_protein_absent      111  anatomy\n",
      "196432        4  gene/protein   anatomy_protein_absent      113  anatomy\n",
      "196433        4  gene/protein   anatomy_protein_absent      116  anatomy\n",
      "\n",
      "[8737 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = gene/protein and y_type = biological_process:\n",
      "        x_index        x_type            relation  y_index              y_type\n",
      "183634      205  gene/protein  bioprocess_protein        0  biological_process\n",
      "183635       21  gene/protein  bioprocess_protein        0  biological_process\n",
      "183636      640  gene/protein  bioprocess_protein        0  biological_process\n",
      "183637      112  gene/protein  bioprocess_protein        0  biological_process\n",
      "183638       36  gene/protein  bioprocess_protein        0  biological_process\n",
      "...         ...           ...                 ...      ...                 ...\n",
      "186671     1421  gene/protein  bioprocess_protein     1783  biological_process\n",
      "186672     1421  gene/protein  bioprocess_protein     1784  biological_process\n",
      "186673     1421  gene/protein  bioprocess_protein     1785  biological_process\n",
      "186674     1421  gene/protein  bioprocess_protein     1786  biological_process\n",
      "186675     1426  gene/protein  bioprocess_protein     1787  biological_process\n",
      "\n",
      "[3042 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = gene/protein and y_type = cellular_component:\n",
      "        x_index        x_type          relation  y_index              y_type\n",
      "182619      112  gene/protein  cellcomp_protein        0  cellular_component\n",
      "182620       22  gene/protein  cellcomp_protein        1  cellular_component\n",
      "182621      223  gene/protein  cellcomp_protein        1  cellular_component\n",
      "182622       75  gene/protein  cellcomp_protein        1  cellular_component\n",
      "182623       11  gene/protein  cellcomp_protein        1  cellular_component\n",
      "...         ...           ...               ...      ...                 ...\n",
      "183629       17  gene/protein  cellcomp_protein      266  cellular_component\n",
      "183630      107  gene/protein  cellcomp_protein      267  cellular_component\n",
      "183631       91  gene/protein  cellcomp_protein      268  cellular_component\n",
      "183632       48  gene/protein  cellcomp_protein      269  cellular_component\n",
      "183633      139  gene/protein  cellcomp_protein      270  cellular_component\n",
      "\n",
      "[1015 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = gene/protein and y_type = disease:\n",
      "        x_index        x_type         relation  y_index   y_type\n",
      "171212      223  gene/protein  disease_protein      543  disease\n",
      "171213       75  gene/protein  disease_protein      543  disease\n",
      "171214       43  gene/protein  disease_protein      543  disease\n",
      "171215      152  gene/protein  disease_protein      543  disease\n",
      "171216       51  gene/protein  disease_protein      543  disease\n",
      "...         ...           ...              ...      ...      ...\n",
      "175123       40  gene/protein  disease_protein     4315  disease\n",
      "175124        4  gene/protein  disease_protein     4304  disease\n",
      "175125       94  gene/protein  disease_protein     5226  disease\n",
      "175126      532  gene/protein  disease_protein     4393  disease\n",
      "175127       96  gene/protein  disease_protein     5227  disease\n",
      "\n",
      "[3916 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = gene/protein and y_type = drug:\n",
      "        x_index        x_type      relation  y_index y_type\n",
      "203540       79  gene/protein  drug_protein        0   drug\n",
      "203541       79  gene/protein  drug_protein        1   drug\n",
      "203542       79  gene/protein  drug_protein        2   drug\n",
      "203543       79  gene/protein  drug_protein        3   drug\n",
      "203544       79  gene/protein  drug_protein        4   drug\n",
      "...         ...           ...           ...      ...    ...\n",
      "205852     4715  gene/protein  drug_protein      776   drug\n",
      "205853       28  gene/protein  drug_protein       66   drug\n",
      "205854       28  gene/protein  drug_protein       67   drug\n",
      "205855      653  gene/protein  drug_protein       66   drug\n",
      "205856      653  gene/protein  drug_protein       67   drug\n",
      "\n",
      "[2317 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = gene/protein and y_type = effect/phenotype:\n",
      "        x_index        x_type           relation  y_index            y_type\n",
      "161392      540  gene/protein  phenotype_protein        5  effect/phenotype\n",
      "161393       94  gene/protein  phenotype_protein        5  effect/phenotype\n",
      "161394       19  gene/protein  phenotype_protein        6  effect/phenotype\n",
      "161395     1207  gene/protein  phenotype_protein        7  effect/phenotype\n",
      "161396     1208  gene/protein  phenotype_protein        7  effect/phenotype\n",
      "...         ...           ...                ...      ...               ...\n",
      "161928     1414  gene/protein  phenotype_protein      112  effect/phenotype\n",
      "161929      108  gene/protein  phenotype_protein      113  effect/phenotype\n",
      "161930      697  gene/protein  phenotype_protein      114  effect/phenotype\n",
      "161931     1409  gene/protein  phenotype_protein      115  effect/phenotype\n",
      "161932     1415  gene/protein  phenotype_protein      116  effect/phenotype\n",
      "\n",
      "[541 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = gene/protein and y_type = exposure:\n",
      "        x_index        x_type          relation  y_index    y_type\n",
      "235494      152  gene/protein  exposure_protein        0  exposure\n",
      "235495      291  gene/protein  exposure_protein        1  exposure\n",
      "235496      152  gene/protein  exposure_protein        2  exposure\n",
      "235497       40  gene/protein  exposure_protein        3  exposure\n",
      "235498       96  gene/protein  exposure_protein        4  exposure\n",
      "...         ...           ...               ...      ...       ...\n",
      "235805       40  gene/protein  exposure_protein       56  exposure\n",
      "235806       19  gene/protein  exposure_protein       57  exposure\n",
      "235807       19  gene/protein  exposure_protein       58  exposure\n",
      "235808       40  gene/protein  exposure_protein       58  exposure\n",
      "235809       40  gene/protein  exposure_protein       59  exposure\n",
      "\n",
      "[316 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = gene/protein and y_type = gene/protein:\n",
      "        x_index        x_type         relation  y_index        y_type\n",
      "0             0  gene/protein  protein_protein     1119  gene/protein\n",
      "1             1  gene/protein  protein_protein        5  gene/protein\n",
      "2             2  gene/protein  protein_protein      120  gene/protein\n",
      "3             2  gene/protein  protein_protein     1436  gene/protein\n",
      "4             3  gene/protein  protein_protein     1437  gene/protein\n",
      "...         ...           ...              ...      ...           ...\n",
      "203535     4575  gene/protein  protein_protein       11  gene/protein\n",
      "203536     4167  gene/protein  protein_protein       14  gene/protein\n",
      "203537     4576  gene/protein  protein_protein       51  gene/protein\n",
      "203538      913  gene/protein  protein_protein      205  gene/protein\n",
      "203539      672  gene/protein  protein_protein       43  gene/protein\n",
      "\n",
      "[14212 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = gene/protein and y_type = molecular_function:\n",
      "        x_index        x_type         relation  y_index              y_type\n",
      "181859       22  gene/protein  molfunc_protein        0  molecular_function\n",
      "181860       22  gene/protein  molfunc_protein        1  molecular_function\n",
      "181861      112  gene/protein  molfunc_protein        1  molecular_function\n",
      "181862       22  gene/protein  molfunc_protein        2  molecular_function\n",
      "181863       11  gene/protein  molfunc_protein        2  molecular_function\n",
      "...         ...           ...              ...      ...                 ...\n",
      "182614     1421  gene/protein  molfunc_protein      366  molecular_function\n",
      "182615     1422  gene/protein  molfunc_protein      366  molecular_function\n",
      "182616     1424  gene/protein  molfunc_protein      366  molecular_function\n",
      "182617     1426  gene/protein  molfunc_protein      366  molecular_function\n",
      "182618      673  gene/protein  molfunc_protein      367  molecular_function\n",
      "\n",
      "[760 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = gene/protein and y_type = pathway:\n",
      "        x_index        x_type         relation  y_index   y_type\n",
      "187217      205  gene/protein  pathway_protein        0  pathway\n",
      "187218      205  gene/protein  pathway_protein        1  pathway\n",
      "187219      205  gene/protein  pathway_protein        2  pathway\n",
      "187220      205  gene/protein  pathway_protein        3  pathway\n",
      "187221      205  gene/protein  pathway_protein        4  pathway\n",
      "...         ...           ...              ...      ...      ...\n",
      "187692       21  gene/protein  pathway_protein      213  pathway\n",
      "187693       21  gene/protein  pathway_protein       11  pathway\n",
      "187694      343  gene/protein  pathway_protein        1  pathway\n",
      "187695      343  gene/protein  pathway_protein      315  pathway\n",
      "187696      343  gene/protein  pathway_protein      316  pathway\n",
      "\n",
      "[480 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = molecular_function and y_type = gene/protein:\n",
      "        x_index              x_type         relation  y_index        y_type\n",
      "230677        0  molecular_function  molfunc_protein       22  gene/protein\n",
      "230678        1  molecular_function  molfunc_protein       22  gene/protein\n",
      "230679        1  molecular_function  molfunc_protein      112  gene/protein\n",
      "230680        2  molecular_function  molfunc_protein       22  gene/protein\n",
      "230681        2  molecular_function  molfunc_protein       11  gene/protein\n",
      "...         ...                 ...              ...      ...           ...\n",
      "231432      366  molecular_function  molfunc_protein     1421  gene/protein\n",
      "231433      366  molecular_function  molfunc_protein     1422  gene/protein\n",
      "231434      366  molecular_function  molfunc_protein     1424  gene/protein\n",
      "231435      366  molecular_function  molfunc_protein     1426  gene/protein\n",
      "231436      367  molecular_function  molfunc_protein      673  gene/protein\n",
      "\n",
      "[760 rows x 5 columns]\n",
      "\n",
      "\n",
      "Sub-DataFrame for x_type = pathway and y_type = gene/protein:\n",
      "        x_index   x_type         relation  y_index        y_type\n",
      "236035        0  pathway  pathway_protein      205  gene/protein\n",
      "236036        1  pathway  pathway_protein      205  gene/protein\n",
      "236037        2  pathway  pathway_protein      205  gene/protein\n",
      "236038        3  pathway  pathway_protein      205  gene/protein\n",
      "236039        4  pathway  pathway_protein      205  gene/protein\n",
      "...         ...      ...              ...      ...           ...\n",
      "236510      213  pathway  pathway_protein       21  gene/protein\n",
      "236511       11  pathway  pathway_protein       21  gene/protein\n",
      "236512        1  pathway  pathway_protein      343  gene/protein\n",
      "236513      315  pathway  pathway_protein      343  gene/protein\n",
      "236514      316  pathway  pathway_protein      343  gene/protein\n",
      "\n",
      "[480 rows x 5 columns]\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "grouped = t3.groupby(['x_type', 'y_type'])\n",
    "\n",
    "# Create a dictionary to store the sub-DataFrames\n",
    "subdfs = {group: subdf for group, subdf in grouped}\n",
    "for key, subdf in subdfs.items():\n",
    "    print(f\"Sub-DataFrame for x_type = {key[0]} and y_type = {key[1]}:\")\n",
    "    print(subdf)\n",
    "    print(\"\\n\")\n",
    "    df_out = subdf[['x_index', 'y_index']]\n",
    "    # rename the columns\n",
    "    df_out.columns = ['src_id', 'dst_id']\n",
    "    # save the sub-DataFrame to a CSV file\n",
    "    #df_out.to_csv(f'data/rgcn/merge/edges_{mapping_dict[key[0]]}_{mapping_dict[key[1]]}.csv', index=False)\n",
    "    df_out.to_csv(f'data/rgcn/pkg/edges_{mapping_dict[key[0]]}_{mapping_dict[key[1]]}.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "test",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
